{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c364b81c-7134-4438-9eb7-74a9da34185b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from uuid import uuid4\n",
    "from itertools import product\n",
    "import concurrent.futures\n",
    "from datetime import datetime\n",
    "import re\n",
    "from collections import namedtuple, Counter\n",
    "from copy import deepcopy\n",
    "\n",
    "from tgsts.sequtils.trimming import trim_intersection\n",
    "from utils import build_full_length_alignments, add_pipes\n",
    "from tgsts.sequtils import str_to_fasta_seqrecord\n",
    "from tgsts.sequtils.kmers import calculate_kmer_distance\n",
    "from tgsts.sequtils.rra import identify_rrs_ali, mask_rrs\n",
    "from tgsts.utils.parallel import run_concurrently\n",
    "from tgsts.typeclasses import MismatchList\n",
    "from tgsts.libs import ANTypingLibs\n",
    "from tgsts.align import mm_profile_from_seqs\n",
    "from tgsts.typeclasses import ExonDict\n",
    "from tgsts.align import exonic_alignment\n",
    "\n",
    "\n",
    "from sfat import Annotator "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25ef7f17-6141-4e16-8995-981ec8973eac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a2672c90-6880-4fee-a032-c89bcb051b24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Library ID</th>\n",
       "      <th>Method</th>\n",
       "      <th>Fragment</th>\n",
       "      <th>NumReads</th>\n",
       "      <th>Fragment Name</th>\n",
       "      <th>cds_mismatch_list</th>\n",
       "      <th>gDNA_mismatch_list</th>\n",
       "      <th>analysis_code</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMAI</td>\n",
       "      <td>ERAP1_BCAv2_11</td>\n",
       "      <td>pbAA</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "      <td>1a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>5utr:-296delACACACACACACAC&gt;, intron1:2595delT&gt;...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAAAGTCCTGGGGGCCACCTCTAACCCACCTTCCTCCTCTACAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMAI</td>\n",
       "      <td>ERAP1_BCAv2_11</td>\n",
       "      <td>pbAA</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>5utr:-331delA&gt;, 5utr:-296delACACACACACACACACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAAAGTCCTGGGGGCCACCTCTAACCCACCTTCCTCCTCTACAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMAI</td>\n",
       "      <td>ERAP1_BCA_22</td>\n",
       "      <td>laa</td>\n",
       "      <td>2</td>\n",
       "      <td>179</td>\n",
       "      <td>2a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>intron5:13741delA, intron11:20063C&gt;A, intron13...</td>\n",
       "      <td>1</td>\n",
       "      <td>TTTGTGTTCCTCTAGTGCTGAGATTTCTTATTCTTGTGGAGGTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMAI</td>\n",
       "      <td>ERAP1_BCA_22</td>\n",
       "      <td>laa</td>\n",
       "      <td>2</td>\n",
       "      <td>213</td>\n",
       "      <td>2a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>intron5:13741delA, intron13:22270insT, intron1...</td>\n",
       "      <td>1</td>\n",
       "      <td>TTTGTGTTCCTCTAGTGCTGAGATTTCTTATTCTTGTGGAGGTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AMAI</td>\n",
       "      <td>ERAP1_BCA_28</td>\n",
       "      <td>laa</td>\n",
       "      <td>3</td>\n",
       "      <td>171</td>\n",
       "      <td>3f</td>\n",
       "      <td>exon15:2285C&gt;G</td>\n",
       "      <td>intron11:20063C&gt;A, intron13:22270insT, intron1...</td>\n",
       "      <td>1</td>\n",
       "      <td>AATGAAACTATAGATAACAATTATTTCTATTATCTTTTCAGGGAGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Samples      Library ID Method  Fragment  NumReads Fragment Name  \\\n",
       "0    AMAI  ERAP1_BCAv2_11   pbAA         1       103            1a   \n",
       "1    AMAI  ERAP1_BCAv2_11   pbAA         1       122            1a   \n",
       "2    AMAI    ERAP1_BCA_22    laa         2       179            2a   \n",
       "3    AMAI    ERAP1_BCA_22    laa         2       213            2a   \n",
       "4    AMAI    ERAP1_BCA_28    laa         3       171            3f   \n",
       "\n",
       "  cds_mismatch_list                                 gDNA_mismatch_list  \\\n",
       "0             No_MM  5utr:-296delACACACACACACAC>, intron1:2595delT>...   \n",
       "1             No_MM  5utr:-331delA>, 5utr:-296delACACACACACACACACAC...   \n",
       "2             No_MM  intron5:13741delA, intron11:20063C>A, intron13...   \n",
       "3             No_MM  intron5:13741delA, intron13:22270insT, intron1...   \n",
       "4    exon15:2285C>G  intron11:20063C>A, intron13:22270insT, intron1...   \n",
       "\n",
       "   analysis_code                                           Sequence  \n",
       "0              1  GAAAGTCCTGGGGGCCACCTCTAACCCACCTTCCTCCTCTACAGCA...  \n",
       "1              1  GAAAGTCCTGGGGGCCACCTCTAACCCACCTTCCTCCTCTACAGCA...  \n",
       "2              1  TTTGTGTTCCTCTAGTGCTGAGATTTCTTATTCTTGTGGAGGTATC...  \n",
       "3              1  TTTGTGTTCCTCTAGTGCTGAGATTTCTTATTCTTGTGGAGGTATC...  \n",
       "4              1  AATGAAACTATAGATAACAATTATTTCTATTATCTTTTCAGGGAGC...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = './erap_tile_test_new.xlsx'\n",
    "\n",
    "df = pd.read_excel(path).replace({np.nan: None})\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8b74be0c-ed29-4e00-8634-dd42f7f8f78d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Library ID</th>\n",
       "      <th>Method</th>\n",
       "      <th>Fragment</th>\n",
       "      <th>NumReads</th>\n",
       "      <th>Fragment Name</th>\n",
       "      <th>cds_mismatch_list</th>\n",
       "      <th>gDNA_mismatch_list</th>\n",
       "      <th>analysis_code</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AMALA</td>\n",
       "      <td>ERAP1_BCA_12</td>\n",
       "      <td>laa</td>\n",
       "      <td>1</td>\n",
       "      <td>393</td>\n",
       "      <td>1a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>5utr:-357delA, 5utr:-323insACAC, intron1:1293d...</td>\n",
       "      <td>2</td>\n",
       "      <td>GAAAGTCCTGGGGGCCACCTCTAACCCACCTTCCTCCTCTACAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AMALA</td>\n",
       "      <td>ERAP1_BCA_22</td>\n",
       "      <td>pbAA</td>\n",
       "      <td>2</td>\n",
       "      <td>77</td>\n",
       "      <td>2a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>intron11:20096delTG</td>\n",
       "      <td>2</td>\n",
       "      <td>TTTGTGTTCCTCTAGTGCTGAGATTTCTTATTCTTGTGGAGGTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AMALA</td>\n",
       "      <td>ERAP1_BCA_28</td>\n",
       "      <td>pbAA</td>\n",
       "      <td>3</td>\n",
       "      <td>715</td>\n",
       "      <td>3a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>intron11:20096delTG, intron14:24183insA, intro...</td>\n",
       "      <td>2</td>\n",
       "      <td>AATGAAACTATAGATAACAATTATTTCTATTATCTTTTCAGGGAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AMALA</td>\n",
       "      <td>ERAP1_BCA_25_reload</td>\n",
       "      <td>pbAA</td>\n",
       "      <td>4</td>\n",
       "      <td>103</td>\n",
       "      <td>4a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>intron14:24484insA, intron18:27923G&gt;A, intron1...</td>\n",
       "      <td>2</td>\n",
       "      <td>GCTTGGCAAAATGTCCTGAAGTCTTGTTGCATAATTTGCTCTCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AMALA</td>\n",
       "      <td>ERAP1_BCA_26</td>\n",
       "      <td>pbAA</td>\n",
       "      <td>5</td>\n",
       "      <td>68</td>\n",
       "      <td>5a</td>\n",
       "      <td>No_MM</td>\n",
       "      <td>intron19:33709A&gt;C, intron19:35500delT, intron1...</td>\n",
       "      <td>2</td>\n",
       "      <td>TCAAGTCAGTTAATACCCTAAGAATTAGATTTTATTTCTTATTCTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Samples           Library ID Method  Fragment  NumReads Fragment Name  \\\n",
       "10   AMALA         ERAP1_BCA_12    laa         1       393            1a   \n",
       "11   AMALA         ERAP1_BCA_22   pbAA         2        77            2a   \n",
       "12   AMALA         ERAP1_BCA_28   pbAA         3       715            3a   \n",
       "13   AMALA  ERAP1_BCA_25_reload   pbAA         4       103            4a   \n",
       "14   AMALA         ERAP1_BCA_26   pbAA         5        68            5a   \n",
       "\n",
       "   cds_mismatch_list                                 gDNA_mismatch_list  \\\n",
       "10             No_MM  5utr:-357delA, 5utr:-323insACAC, intron1:1293d...   \n",
       "11             No_MM                                intron11:20096delTG   \n",
       "12             No_MM  intron11:20096delTG, intron14:24183insA, intro...   \n",
       "13             No_MM  intron14:24484insA, intron18:27923G>A, intron1...   \n",
       "14             No_MM  intron19:33709A>C, intron19:35500delT, intron1...   \n",
       "\n",
       "    analysis_code                                           Sequence  \n",
       "10              2  GAAAGTCCTGGGGGCCACCTCTAACCCACCTTCCTCCTCTACAGCA...  \n",
       "11              2  TTTGTGTTCCTCTAGTGCTGAGATTTCTTATTCTTGTGGAGGTATC...  \n",
       "12              2  AATGAAACTATAGATAACAATTATTTCTATTATCTTTTCAGGGAGC...  \n",
       "13              2  GCTTGGCAAAATGTCCTGAAGTCTTGTTGCATAATTTGCTCTCAAA...  \n",
       "14              2  TCAAGTCAGTTAATACCCTAAGAATTAGATTTTATTTCTTATTCTG...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Samples']=='AMALA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e9c922f5-d0a4-4d41-ae57-9b9931fb3a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Fill rows\n",
    "rows = []\n",
    "\n",
    "prev_filled_row = None\n",
    "for i, row in enumerate(df.to_dict(orient=\"records\")):\n",
    "    \n",
    "    if row['Samples'] is not None:\n",
    "        prev_filled_row = row\n",
    "        \n",
    "    for column, value in row.items():\n",
    "        if 'Typing' not in column and value is None:\n",
    "            row[column] = prev_filled_row[column]\n",
    "\n",
    "        \n",
    "    rows.append(row)\n",
    "    \n",
    "\n",
    "#Filter rows\n",
    "\n",
    "#Get nonzero analysis codes and remove missing seqs\n",
    "rows = [\n",
    "    row \n",
    "    for row\n",
    "    in rows\n",
    "    if row['analysis_code'] != 0\n",
    "    and row['Sequence'] is not None\n",
    "]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e82dae36-ecd4-4549-ae0a-954d3784f24e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b5a14ab-0a51-424a-b64a-a871b717529b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e6f9511e-98fa-4863-9134-4dd087495cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "erap_reference = 'GTACAGTGGCCCTTGGTAGTGCAGGAAAGTCCTGGGGGCCACCTCTAACCCACCTTCCTCCTCTACAGCATCTCCCACTGTAGTCATTCTCTACCGAAGCCCCAGAAGGTGCGGCACTTTGCCACGACAGAGTACTGGGTTCATGTTTCTTTCCGAGGCGGGCCAAGAGCTCTCAGCCCACTGGCAGTGGCGAGATGACGGACACCCAGCGAGTCCAATGGGCGTCGAACGCGTCTAGGCTTGGTGGACTTGTCAGCGCCTGCCTGGCTTCGGTCCCCAACTTGAGCACCGGCCCTTTCCTGCATGCCCCTAACCCTCGCAACGCTAAACAGTGAAAAAAAAAAAAAGACAAAAACAAAAAGCATCTCAACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACGGATCCGCGTTCAGAAAGGCGTGCACTTCCTACGCCTGATCCCCCGCATCGCAACCTCGCAGCTTCCCCGGCGTGCAGCGCTCATTTACCAATTCCCTTCCTGGGAGTTGCGGCTTCCCTCGCTCGGCCCCACTCCCGTTTACCCTTTCCCCAGCTCCCGCCTTAGCCAGGGGCTTCCCCGCCTGCCGCTAGGGCTCGGGCCGAAGCGCCGCTCAGCGCCAGCCTGCCGCTCCCCGGGCTCCACTTTC|ACTTTCGGTCCTGGGGGAGCTAGGCCGGCGGCAGTGGTGGTGGCGGCGGCGCAAGGGTGAGGGCGGCCCCAGAACCCCAG|GTACAGCGCGCTCGAGCCGCGGGTAGGGGACTGCGGGCCGGGAGGAGAGCGCGGCACCCGCCCCTTCCCTGCGCCCGTCAAGTGGGGGGCTGAGGGCCTGGGGCACGGGAGGAGGGAGACGGGGCACGGGAGGAGGGAGACGGGGCGCGGGAGGAAGGCGACGGGGCGCGGGAGGAAGGCGACGGGGCGCGAGACAGGGCGCGGGAAGGGCGGGGGGAGTCGCTGGCTAGGCCCGAGTCCGCGGGGTGCCCGGCGGGTTGGCGGCGGGCCCACCCCTGCCGGTCCCTGTCCCTGTCCCTCCGGGCGCGTGGCCGGTGCGCCTGCTTCACGGGTCTCCCCGCTGTTCGGCCGGCGGGAGCCTCCCTCAGCGCTCCGCCTGGCGCCTGGATGCCTGCCAGTCCTGCAGGCCACCGACGCCCGCGCGAGGCCAAAAGGCGGGGTGGGGCGGGCAGCTGGCTCGGGCTGAGGAGGGCACCTGCCCATAGCTGCTAGAGAAACCCAGAGGCTTTGGGTTAAAGACTCTGGTGGGGTGGGATGCGCGGGCCGTGTGTGTTCTTAAGGTCACTTCCCTCCCTGCTTCTCCTGTTCTTCTGGTCAGCAATTCTCTCTCTCCCCTTCGCTCTGGCTCTGGCTGGGTTTTATTCAGATAAAGCACCTCTGTTGACGCAAATTAAAAGTTTCCTATCTGGGTGCCTCACTGGCCAGGTGGTCCTACAAAGTTAATTCCATGAGGGGAAGGGGGAGAGCACACACTTCCTCACGCTTTTGGATTTCTTTGTGTAGGCTAGGTTCAGAAAGAAATTATCTGTTTCCTATTAAACACCCAGAGGATTCGCTCTGAACTCAGGACGTGGTCAACAATTAACAAAACAACAAAACAAAACAAAACAAAACAAAAAACTTGAAAATTGGGCACAGTTGTCTCTTGCCTGAGGATTTTTAATTAGTATAAGTAGCACATTTTCAGGTGCGGCCTGAATAGAAACATTCTAGTACTTTTTTTTTTTTCAAATTAATCCAGCATTTTTATTATTTACCAACAGTGCTTGTTAATTTCATTGTTCAGGAAATTCTGGAAGAACCTCAATTACTTCTTGATGATCTATTTCATATACTATAGTGCCCCAATAAAAGGAAGGGAGGCAGAGGTTGCAGTGAGCCAAGATCGCACCACTGCACTCCAGCCTGGGAGACAGAGCGAGACTCCGTCTCAAAAAAAAAAAAAAAAAAAAAAAAGAGGGATCTTGGTGAAGCTGAAAAAGCAGGGAGTTTATACTCACACAGATGTGGATTGCATTCCAACAAGTTGTGTGAACTTAGCAAAGTTACATGAATGGTTCTTTGCTTCAGTTACCCCACCTTGAATGAGAATAATAGGCTATTAGAGAGAGTGAGCATGTGTAAAGTGCCTGGCACTTTGGAGGAGCTTAGTAAATATTAGTTCTTTTTCCTCCTTGGCTTCTATTCCTTTAAGTGTTAGTGGAGTGTAACTTCAAAGAGAATTGTACTTCTCTTGGGAAGAGCTGCTTATATTGAGTAGACTACTGTTTTTGAGAGCTTTCTTTTTTTTTAACTGCCTCCCTTTCAGAAAATTGTTAGTAAAACCAGACCCTAGGAGACCAGCCAGAACCATGAAATGCCATGTTTCAAACTGGAACACTTACTATTGGAAGCAAAGCCAAACAAGAGCTAATTTTTCAGGGGAAGAGCCAAACACACGAACATGATCACAGAATCTTGGATGTAGCCTACAGTTTGGGATTAATAAGAGAATTTATTAGTGAAGCCCTTTATTCACTACATGGAGTTTTTACCAAGCCCCACTCATGCACTGCATCCTCGTTGAGACATAACTGTTTCTCTTTGGACCCCTCATGGACCCAACCCTGCAAAGCCTCTGATCCAAGGTCCCGGTACCAACCCCTTCCGCAGCACATCAGCCTTTCTGTCAGCTCATAACGAGTTGGAATTTCTAGATCTTCTCTGGGGCTGTTGGAGAGGTCTCGGGGACTTTCAGAGTCCTTACATGCTTGAACCTGCCACCTTCACAGAGTCCTCTGGATCCCGTCTTGGGGCAGGGGCGATGCTCATTAAGCTGTTGCTGCCAGTAATTCCATATGGAAAGCAAAACACAAGTTCCATTTACTCTCTAGTTCCCCAACTTCAAGGGCAAAAAAATGTTCTCCCTGTTCACACTTCCTGTCTCACCTGGGTGGTGCCTTTTGAACTGGGATTATGAGATTTCCAAGACTCTCTCTAATGTGTAGGTATCCTTTCTGTTTAGCCTCCAGATTGCTCCAGAGGTGAGGAGAAGGGAATTCCCTTGAGCTGTGCATTTGGGAAGGGAGCAAGGAAGTCAGGGGTTAGGGAAGGCACTTCAGCCATTGCCTTGAATTAGTATCCTATCACATAGAGTTGAAGGGGGAAAGCCAGGATTTGGCAAGGATGAGCTTTTCAACCTTGGCTTCTCAGTAAAATCTCTGGACAGTTTTTTTTTTTTTTAAAAAAAAAAACCAAAAACCAGAAATTAAACAATCCGTAAAACCATACCCCGATCTTATCACCATAGGTTCCAGTTTAATTGTTGTCAATAAGGACCCAGGCATCGAAAATTTTAAAAGCTTCCCAGGTTACATTAATATGCAGCCAGAGTTAGGAAACTTGACATCTCAGAGAGGGAGAATTCCATGTACTGTGAACACTTTGAGGGATCCACATTGTAAGCTTGCTACTTTTCCCAACTGGAACACGAGGAGTTTGGGCCAATCACTTGCATTCACCTGGTATGGGCCCCTCTTGGCAAACACCCATTAGAAAGGTGCGTTTGTATAAAAGAAATAAAAACTTATGTTTGATGCTTGGGGCATGGTTTGCCAACTTCCTTAAAATTCACATTGCCTTTTTTTTTTTTTTTTAAAGACAGAGTCTTGCTCTGTTGCCAGGCTGGAGTGCAGTGGTGCAATCTCAGCTCACTGCAACTCTGCCTCCCAGGTTCAAGCGATTCTCCTGCCTCAGCCTCCCAAGTAATTAGGACTACAGGTGCGCACCACCACACCCAGCTAATTTTTGTATTTTTAGTAGAGACGGGATTTCGCCATGTTGGCCAGGATGGTCTCAATCTGTTGACCTTGTGATCTGCCCACCTCATCCTCCCAAAGTGCTGGGATTACAGGCGTGAGCCACCATGCCTGGCCTGAAATTCACATTGCTTTTATGTCTTTAAAATCAATCCAGGTGATTGATACGTTTGCCACAAACTACTGGGAAAATTAAGCTCTTTAAGCTCTTCCTGGCTAAATAAATAGGTAATTAACTTTTGGCATGACAATTTGAGGAAGACACTGATGTTATTAAAGGTTACCATTACATATACTTACAGGGGAAGTGGAGACCTCGTTCTAGTGGTGGCTGCCACTGTGTCAGCATTTGATTTCCAGGGTACTAGGGTGACTTTCTCCAAGGTCCAGTGTCAGTGGGGAGTGGTGCTTGATCAGATGTTCCTCTGATATGGTTCTGGTTTTGCTTCTTTCATGAGCCTGGTTTCCCAAATGTCCTGCAACTCTGTGACCCGTGTAGTGAGCCACTCAGGATCCCCTAATGATTCCTTTTCTGCCTATATCAGCCAGAGCTTGTTCGGTTGCTTTCAACCAAGAACCCTGACAGGTAGAGTTAATTTAAACTTTGAACATCAAATGATGCTTTCTAGTACGTGTTAATGATTGTTGCTAACTGTAAACATCTTTCTTATATGAAACCATAACATAGGGAAGGTCCTTTTACTTTCAGGAAAAGACCTAGTACTTTTGGAAGTTTATGCCTATTTCTGTGAATGCTGGGTGGATACATTCTGAAATTATGCTGTGTCAATAACATTTTAATGACATATATTTTTGCTTTTGTACATTTGTGCCGCTAG|GTAGGTAGAGCAAGAAGATGGTGTTTCTGCCCCTCAAATGGTCCCTTGCAACCATGTCATTTCTACTTTCCTCACTGTTGGCTCTCTTAACTGTGTCCACTCCTTCATGGTGTCAGAGCACTGAAGCATCTCCAAAACGTAGTGATGGGACACCATTTCCTTGGAATAAAATACGACTTCCTGAGTACGTCATCCCAGTTCATTATGATCTCTTGATCCATGCAAACCTTACCACGCTGACCTTCTGGGGAACCACGAAAGTAGAAATCACAGCCAGTCAGCCCACCAGCACCATCATCCTGCATAGTCACCACCTGCAGATATCTAGGGCCACCCTCAGGAAGGGAGCTGGAGAGAGGCTATCGGAAGAACCCCTGCAGGTCCTGGAACACCCCCGTCAGGAGCAAATTGCACTGCTGGCTCCCGAGCCCCTCCTTGTCGGGCTCCCGTACACAGTTGTCATTCACTATGCTGGCAATCTTTCGGAGACTTTCCACGGATTTTACAAAAGCACCTACAGAACCAAGGAAGGGGAACTGAG|GTATTTTTTTTTCTCTTTTTCTTTTAAACTGCAAGTGCTGCCCACGCTAAATTCATTATTTCAGATTGATTGTCTTTTAAAATTCCCTTTGCTGTTGAACTTTTTCTTCAGTTTTGCTTTTGCATCTTCTTTATAGTGTTAAAAATGGCTTTTTCCCTTGCTTTTTAAATCTCATTTTAAAATTCTATTTTAACCAATTTTCTTTCCCCCAGCTCTATCAGAGTAAATATCTATTTGTTTATTTGGTTCGATTTCTGAGACATAATAAACATGTTTAATTTTCCTGAACTGTGTATTAGTTTCCTAGGGCTGTTGTAACAAAGTACCACAGACTGGGTAGCTATAAACAACAAAATGTATTCTCTCTCAGGTCTGGAGGCTAGAAGTCTGAAATCAAGGTGTCAGCAGGCCCGTGCTCCCTCCAGACTCTGGGTAGAATCCTTCCTTATATCTTCCTAGCATCTAGTGGTGGCCGTGGATCCCTGGTACTCCATGCCTAGCACCTGCGTCATTCTAGTTTCTCCCTCTGTTAGTCGCATGGCCATTCTATTTTCCTATGTCCCAGTCTCCATCTTCTTATAAGGAAACCAGACATACCAGATTAGGGCCCAGCCTGGTGGCCTCTTCTTCACTCCATGATACTTGTAAAGACCCTATGTCCAAATAAGGTCACATGCACAGATACTACAGGTTAGGACTTCAGCAAATCCTACTAGCAACATATTAGAGGAAATACTTTTATCTCAGTTAAAACTTTTTTAGAGATCTCTTCTCACCTTGCTTTGGTTCTGTTTTTAAGGAGGAGACTTATTTGGGGGAGATTTTATGCTCAGTTTTAAAATGGAATTTTATTTGTTGGTAGATTATACTAATTTATTTTTCAAATTCCATATTATTTTATCAAGGTAAGAAAGTAAAATTTATTTCACTCATAGCCCCCTGAACTGACCACTACTTCTATTTCACTGGAATATTCTGCCAGACTTCTTTCTGGATATGCATATATATTATTTTATATAAATAGAATTCTTATATTTTCTCTTTTACAAGGAAATTTTTTAACTTAATATGTTGGGAAGATGTTTTATATAAATGACTATTACTAGACATCTTTTTAATGATTTTATTAAAATACATAGTTGTACATTAATATCTTTAATCCCTCGGGAATGGAAATGAAATTGTTTCCAATTTTTCAGTATCAACAACACTTTGATGAGCATTTTTGTAAATTATTTCTTTAGAATAAATGCCTAGAAGTAAAATTGGTAAGCCAAAGAGCCTATAAATTTTTGATAAACTTTGTCATTTTTTCTTACCCCTCACCAATAGTGTATATTGGCAGTCTTTTTAGAGCAAAAAAGGTCCCCAAATAAATGATCTCTTTTTAATTTACGTAACTTTGATTAAGGAAGTTGAGCAGCTTTTGATATGTTTGTTAATGATTTATAGGTTCTTCTTTTATGAATTGCCTGTTAATGACCTTTGACTTTGCCTGAGATTCCCTTGGTTGTTGGGGATTTTTTTTTTTTTTTTTTTTTTTTTTTTGTCGAGACGGAGTCTTGCTCTGTCACCCAGGCTGGAGTGCAGTGGCGCGAACTCGGCTCACCGCAACCTCCGCCTCCCGGGTTCAAGCAATTCTCCTGCCTCAGCCTCCTGAGTAGCTGGGATTACAGACGCGCACCACCACACCCAGCTAATTTTTGTATTTTTAGTAGAGATGGGGTTTCACCATGTTGGCCAGGGTGGTCTCGAACTCCTGTCCTCGTGATCCGCCCGCCTAGGCCTCTCAAAGTGCCAAGATTACAGGTATGAGCCATTTTTAGGATTACATATTTTTAGGATCTCACTATCTGTTAAGGCTATTAAGTTTTCCTCACATACTTCTTAAATGAATTTTCCCTGTTTTTTACTTTTTCTTTTTAACTTTTTTTTTTTTCTTCCCGAGACAGGGTCTGGTTCTGTCGCCCAGGCTGGAGTGCAATGGCGCAATCTCAGGTCACTGAAACCTCTGCCTCCTGGGCTCAAACCATCGTCTCACCTCTGCCTCCCAAGTAGCTGGGACTACAGGCCTGCACCACCATGCCTGGCTAATTTTTGTATTTTTGGTAGAGATGGGGTTTTACCATGTTGCCCAGGTAGGTCTCACACTCCTGGGCTCAAGTAATCCTCCCACCTCAACCTCCCAAAAATGTGTTAGGATTACAGGCATGAGACACCATGACCATGCCCAGTGTAACTCGTCTTTTAACTTAGTTTATGATAGCTTTTGTCAGATGAAATATTTTTTAGTAGCAAAATCCATCAATCTTATAGTAACTTTAGGAATTAAAAAGCAAACACCTGTTTTAGAGTTCCTGGTGCTTCATGCTCAATTTTTATTTCACTTGCCTTAATTTTAG|GATACTAGCATCAACACAATTTGAACCCACTGCAGCTAGAATGGCCTTTCCCTGCTTTGATGAACCTGCCTTCAAAGCAAGTTTCTCAATCAAAATTAGAAGAGAGCCAAGGCACCTAGCCATCTCCAATATGCCATTG|GTGAGTCTGCACTCCTGTGTATTTTCTATAGGAAAATCTACTGATTCTCTGTGACTTGCACTAGCCCAGTGACAGTCAACATTGGGTCACCTGTTTTGTTTTATTGCCTGGCAGATCGTTACTAACTTTTCATTTATAACCTATGCTTTTGTTTCAAGCCATAGTTATATGTAATCAAAGTAAAAATTGCACCTAAAAATGCAAGATTTCAGTAACAGTGCCATTCCAGGTTATACATGCTGATAGGAGGGAAGTGGTATAAGAAATTCAGGTCAGGTTTAAATATTAGTGCCCTTCACAAAGCACTTTCACCCTCATTTTCTCATATGATCTTTTAAAATGATTTTAAAGGTAATGCTTTGTCAATAAGGCCATATTTTTAGCATACAGTTATTTTCTCTAAGTTACATATACTATATAGTAATATTCATTATATAATTTGTAACTGAATGTACAAAATTGGGCAGACAGAAAAAGAGAATAAAAGTAATCTTTTCCAAATATTATGGTGCTGAAGGTAAGTCATGATAGATAGCTTAGCTTCCAGAGGGAAACTATTATTCCCCAATCTCAATGCGGATGTGGACAGCATTCCCTCTGATTTTTAAAAGTGACTAGAAGATGACCATGCCAAATGAATAAAACTGTTCAGTAAGTGCCATCATCCTTTGATTCTGGTAGTTTAGAAAAGCATCAGCTGGGCCGTCATTCTGCAGCTGGTATATAACACCTCCTGGAAGCACATCCTTTGTTCAGAGAAACTCACTGGGGATCAGAGTCAGAGTAGAATAGGCTTTGCCTAGAGTCCTGAGGGAAGAACAGCTTTGTCCCTGTGCTGACCGGGGAAGCAATATCATAACATGGAGAGATACTGAGAGCCACAGACCAACCTCTAGTGTGGTGCTTCTCAACCTTAAAACTCTATACCTACCTGCCTTCATACCATAAGGATGCCTTCCTCAAACAAGATTCGGATCTCCCACCCCACGCCAGAGGGCTATCCTCTGTAAAAATCACTCTTCTGCAAATTCCTACCAGCCAAGAACTTCTGTCCCCACTCCCACCCTTGTATATGAAAAGACAAGAAACAATTATGCTATTTTCCTAATATAAATTTAATATACAGGATGTGCTTTTTCAAAATATAGTCCTCTCCCCCAGTATTGTGGTATTACTCCCTGGGACAGAAGTGAGTTCTTTAATTGGTGAATCAAAGTTCTAGGAGAATAAAGGACCAGGGATGGGAAGGAGACAGGAGAGAGACTGAAGGACCAAACAGGATTAGTGGAGAAATTTGTAGGCTTTCAGAAGGGAGGCCTGGAGCTTTGGAAGCGCCACAAAGATGCTACAGTCTAAATCCATGGATATCCAGGATCCACTTAGTGAAGATAGGAAAACTTCTTTTTTTTTTTTGAGGATAGGAAAACTTCTAATGCAATGTTGTCCCTTTGGACGGGAATACCTCTCACTAACAGAAATCTAATACGAGTAGCCTGACCTCAGGCTGCAGATATTGAGCTGAGGGGAGAACAATGGGGTCTCAAAAGATCTTTTTGGAGACCAGAAAAACACAATATATACCATTGGAACATTGAAGCTTTTGGGCATGGGGCAGAAATTAATCACATTTAAATTTGAATTAATTTAATCAGGTTATTTTCCTAATAATTAACACAACTCGAGAATGGAAATTTTTGGCCAGGTGTGGTGGCTCATGACTGTAATCTCGGCACTTTGGGAGGCTGAGGCAGGTGGATAACCTGAGGTCAGGAGTTCAAGACCAGCCTGGCCAACATGGTAAAACCCTGTCTCTACAAAAATACAAAATTAGCTGGGCGTGGTGGCACATGTCTGTAATCTTAGCTACTTGGGGGGCTGAGGCAGGAGAGTCGCTTGAACTCTGGAGGTGGAGGTTGCAGTGAGTCAAGATTGTGCCATTGCACTCTAGCCTGGGTGACAGAGTGAGACTCCATCTCAAAAAAAAAAAAAAGGAAATTTTTGTTGTAGGTAGGCAGAAGCAGAATGCATTTAAAAAGAAAAGATGATTTGGGATCCTTTATGAGTAATCCTAGGCTGGGTAGCAGAGTTGGTTTGAATGACCAAATAGTGACCAGAAGTTGGTGGCTGATGGGTATTAAGAAGGATGAGGGCCAGGTGAGGTGGCTTATGGTTGTAATCCTAACACTTTGGGAGGCAGAAGAAGAGGATTTCTTGAGGTCAGGAGTCCAAGACCAGCCAGGGCAACATAGCAAGACCCTATCTCTCAAAACAAAAAAAAAAGATGAGGTCAGAGCAATAGAGGTAAGTATTGGATTACAGGAAAAATGCCCGTGACCATGGTTTCACCCAGCTAATTCTGGCTGGTTCTTTTTCCATCTCCGTGCTTTTTATTGCTGACGTGTTAGACTTTCTTCTTTAGGGGCAGACCTCTAAGACTGTACCTCCATCAACTATACCCCACCCTTACTCTCTGATTGCACTTAAAAAGGTGATTCCAATGAAGCAAATGAAGCAAATCTTTTTTTTTTTTTTTTTTTGAGATGGAGTCTCGCTCTGTCATCCAGGCTGGAGTGCAGTGGCGTGATCTCGGCTCACTGCAAGCTCTGCCTTCCGGGTTCATGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCACCTGCCACCACGCCCGGCTAATTTTTCATATTTTTAGTAGAGATGGGGTTCCACCGTGTTAGCCAGGATGGTCTCAATCTCCTTACCTTGTGATCCACCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAGGTGTGAGCCACTGTGTCCGGCCCAGGTTACTTTCAGTTATACAGCAGAACAGAAGCTCTTTTAGGTACTACAGGGTTATATATTTTCCCCGTTGCGTATATGCTCAACAGCTCGACATTGCATTGCCAGATAATTCTCAAACCTGTATTTAAGGAAAAGTGGATCAGCCACATCTTGGCAAAACTCACAATTTCAGTTTTGCTTTGTCTCATCCGTGTTATCAATCCACATATGCCAAATGTGGATTTACAGTGTATTGTAAACTTTAAAATGGTAAGTTGTATGGTATATGAATTATATCTCAATAAAAAAGAAATTGAGTAGAACTGTTTGACGTTAATGTCTAAATTATAATTAGACATTGGAAAGATAACTTTTAAAGTAACTATAGAAGCGTCATTAGACAGGGTCTGGCTCTGTCATCCAGGCAGGAGTGCAGTGGCTCAATCTTGGCTCACTGCAACCTCCACCTCCCAGGCTCAAGCCATCCTTCCACCTCAGCCTCCCCAGTAGCTATGACTACAGGCACGCACCACCAGCAGGACTAATTTTTGTATTTTTTTTGTAGAGATAGGGTTTCACCATGTTGCTTAGGCTGGTCGCAAACTCCTGAGCTCAAGCATTCTGCCTACCTCGGACTCCCCAAGTGCTGGGATTGATACACTTTTAATATTATGTCTGATAATTAGGAAATTTATCATGTTCACTGTATTGGATAATTGGATTACTTGATAATTTGAATTATTCTGATTTTAG|GTGAAATCTGTGACTGTTGCTGAAGGACTCATAGAAGACCATTTTGATGTCACTGTGAAGATGAGCACCTATCTGGTGGCCTTCATCATTTCAGATTTTGAGTCTGTCAGCAAGATAACCAAGAGTGGAGTCAAG|GTGAGCCTATGACTGTCACATATGGTGACCAGCTTGTTCTGGTTTGCTTGGAACTGGTTTTAAAACTGGAAGTCTGCCTGAGCGCAGTGGGTCGTGCGTGTAAACCCAACATAAACCCAACAGTTTGGGAGGCTGAGGTGGAAGAATCACTTGAGGCCAGGGGTTTGAGACCAGCCTGGACAAAATAGTGAGAACCTGTCTCTGCAAAAAATAAAATAAAAAAATTAGCCAGGCATGGTTCCTTGTGCCTGCTACTAGTCCTAGCTACTAGGGAGGATCCCTTGAGCCCAGGAGTTTGAGGCTTCAATGAGGTATGATTGTGCACTCCAGCCTGGGCAACAGAGCAAAACCATGTCTCTAAACAAACAAACAAAGACAAAACCAAATACCAAAATCCTGGAAGTCCTGCATCCTGGGAACCTTCTCAATCTCAGGCAAACTGGGATGGTTGGCCAGCCTGTTGTCACGGATGCTCATTTGTATAGTGAGGTTCTAATAACAACAACGTGGAGAGAGTGTGGCCTGGCCTGAGTCATGATCCTGGCTTCACTGCAGTCACTTCACTGACTCTCTGACCTTGGCCCTATTCCCTCTGAAACTTAGTATTTACTTCTTTGGAAGGTATAACTTGGACTAGATCCTGCAATGGTCTCTAAGGTTGCTTCTGGTTATGGCTTTCTGCAGTTTGGAAGTAAATGTTACTATCTGGCAGGGGATTTCTGGCTATGGTAAGGAAGATAGAGCAACCTGCTTGGAATACCCAAAGGCTTTGGGCCAGGTACACTGGAATGCTGGAGAGAAAAATCTTGTTTCAAGGCACACTTGTTCCTCATTTGGGTACTGTTGCATAGTGGGCAACCTATTCAACTGTGTGCCGTAGCTCAGAATGCAAACAGGTTTTTCTGAGGGGAGGAAGGGATGCTTTGTTTGAAGATACCTTATGTGTTTGTGCTGGTTTTCACTGAGGCCTGAATAGATGGGGATTCCCTGCTGAATTGCTTTGTGTTCCTCTAGTGCTGAGATTTCTTATTCTTGTGGAGGTATCTTTACTTAACTGGGGATTTGAAGGTGACGCTTGAGACTCGGATGAAGGGAACATTCTTAATTCAGCAGTGAAACTATCAGCAAAAACACCCGCCCATTGCTTTGCCACTTATCTGAATCTCTTAGAAATGATTATTTTAGTAATGTCTAATCTATATTAATATTTTTAATTCTTTCATTTCTTTAAACACATTAAGCATACAATTATATATCTGTGTCTGGTAATTGTTTTATCTGAATTCTTTGTGTATCTGATTTTGTGGTTCGTTGTTTCTGCTGGCTCTTGCTTATGGTATCTTGTTTCCTTGTTTGTATTATGAATTATGTTTGTGAGCTTACGTTGCCTGAGTCTAAAGTGGATTATTCCAGAGAGAAATTGTATTTGCTCTTACAGAGTGTCTGGGAGTACTTACTGGTCCAGGGATCACTTTACTTGTAGTTTCCTTGAGAAAGGGTAGTTATTTCTAGTTTACCTTTACATTAAAGGCCTGGCCTTTGGGTACTAGCTTTATGCAGGGATTGTATGTCCTGTTAGACTTTCTACTTTGGGCAGGCCCTGGACTTGGTCTCTTAACTCCTGAGTCCTTCAATGACATAAGAACCAAAGCTCAAGTCCAGCTGTGTTGGGCTAGTGCCGGCAGGGTTAAAGCTGGCTGCAGTGCTCTCCTGACATCAGAGGGTCTAACTGTCATTTCACTTTGGCTTCTAAATCTTTCTTTCTCATTTGCCATCTTATAAACACATTTAAGAACACTTTATACATGTTATCCAGCATTTGTTGTTGTTTTCAGAAGGGGGATTAATCAGGAACAGTCAGTATTAATGCAAGAAATGGAATTCCCAATTATTTTCTTTAATATTGGCAACCATATCCCACAATATGAAGACATTAATGTCAGTCTTCTACACAATGTGGGGAGAGAAGCCAGTTAAGATATTTGAATTCCTTTCTGTGCCTTTCTCTTTAG|GTTTCTGTTTATGCTGTGCCAGACAAGATAAATCAAGCAGATTATGCACTGGATGCTGCGGTGACTCTTCTAGAATTTTATGAGGATTATTTCAGCATACCGTATCCCCTACCCAAACAAG|GTAGAGATTTTGCACAGATATTACACATGACATTTGATGAACACAGTCATAGATTTGTCATTATAATTGGCACATCCCTGTAGTTGCCTCAGCAGCCCCTCAAGCCACAAAAACCCCAGCAAGTGACAAACCTGCGGTTGATCTTTCTGAGCATCTCCTCACCCTTGATGAGTACAGTAACTTCTAGTGATAGTGAAGAAAGCAGATCTTCATAGAGTTCTTGAGGCATATGGCATGGGGACTCTTTTGCCTTCTGATTTTATTAGTGGGCAGACAGCAGAGGGAAGAGGCTACATTTTTTCTTTACTGGCACCTGCTTGGCAGGAACCCAGAGGATGCTCAACAAACTGTTTTGAATGAATAAATTTATATAGTGTAAGACAATCTGAATTTTCTTTCTTACACAAGCCCTTAAAACTATTGTATGTTTGATTTTTTAGGTATATGTGGGTCTTGGGCATCCAAAATAGAATGGATTATTATAATTGTTTAGTTTTTCAATTTCCAAACTCCTCATCAGAAGGTTAGAAATGGAGTCAAGAGGCCTGAAAAGGCGGGCACGGTGGCTCATGTCTTGTAATCTCAGCACTTTGGGAGGCCCAGGCAGGTGGATTACTTGAGCCCAGGAGTTTGAGGCCAGCCTGGGCAACATGGCAAAACCCTGTTTCCACAAAAGTACAAAAATATTAGCTGAGTGTGGAGGTGCACTCTTGTAGTCCCAGCTACTTGCAAGGCTGAGATGGGAGGATCACCTGAGCTTGGGAGGTTGAGGCTGCAGTGAGGTGTAATTATACCACTGCAAATGCACTCCAGTATGGGTGACAGAGTGGGACCTTGTTTCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAGGCCTGAGATGCAAGTCTGACTTTGCCACATTTATAGCAAAGTGATGCTGAGTCACTAAGCCTCTTTTTCTCCATTGTGAAAGGTTCTTCCAGTCTAGAGCTCCATGATTATGCACGGGCAGCTGGTTCAAACACCTATCCATTCTGAAGATTAGTGTTTGGGAGAATGTATAGCTTAGAGACTGGTAATGTATTTTATTACTTCCTTCCCAAG|ATCTTGCTGCTATTCCCGACTTTCAGTCTGGTGCTATGGAAAACTGGGGACTGACAACATATAGAGAATCTGCTCTGTTGTTTGATGCAGAAAAGTCTTCTGCATCAAGTAAGCTTGGCATCACAATGACTGTGGCCCATGAACTGGCTCACCAG|GTATAAGCTCATTCACACTTTTAATAAAGTATAAACTACATTTATATTGCTTCTATGGGACATATAAGGCTATTTATATAATTTTTACTTTGTCTTTTTTTAATAGGAAAAATTGTTTCTCCAAAGCATTCGTTTTTATGTCTTATAATGCATGTTGAACTTTTTTTATTTTTACCTTGATTAAATATTGGTCCTGTAAATATATGTTAACATTCATAAACTTATATTGGACATCTAAAATATACTTCTTTCTGAGTGTCTTTGTTTATGGCTTATGTTGTGCTTTTAG|TGGTTTGGGAACCTGGTCACTATGGAATGGTGGAATGATCTTTGGCTAAATGAAGGATTTGCCAAATTTATGGAGTTTGTGTCTGTCAGTGTGACCCATCCTGAACTGAAAGTT|GTAAGTAGTTATTTATCCTTCACATTTGAGGTTAATTTGTTGTTTTGTTCAATATTGCTGGAAAATATTCACTAATCTTTGATTATAGAACTTATAAAAATATTTCACTGATAACTTCCTTGACAGTTTAGATATGAATCGTGTTGCAAAAATGCTAGTGAACTTTCAAGATATACCACAAAGACTTATTTACAAACTCAGTTGTGAATCTATGATTGTTTGTAAATTGTCCGATTTTATATTTCTTAATATCAAAGAAATAGAAATAATGGCAGCCTAAATGTTCCATTTCATTTTCCCAACCTTCAGGTTGCTCATTGCAGAATTATTAGGTACAGATTACTGATATCTCAATAGGACCCTACTTGCCAAATCAAATGAGTTATACTTAAGGTAACTGCACATTTGATTGTATAACAACCTAGGCTTCTGGATCACAAATTTATTGCTTTGGAACTGCTATATGGTTTATTTTTTAAATCACCACATTTAACTTAAGAAAATCACATATAGCAAAATTAAACTTATGTAAAAATTTTTTTATCCTACTCCAATGATTCTTTTCATTTGTAAATATTTATTCCTATTTTCATCTGCCTGTATACATTATTTAAAATATACATTTGTGAATATTTATTACTGCCTATTTTCATCTGCCTGTGTACATAATTAGAAAGTACACATCAATATTGCATTATAATCCTAAATATTTTCTTATGTTTCTACATGATCTTTAGTAATAAAAATGATTATGGATACATATTGTCCTGTTGAGCTAATGTGCTCTAATAAAGCTGGTTAAGGTTCTCCATTTATTTTCATTATGTTTTTAAAGGTAAGTATCTTTGCTCTAAAACAGTACAGACAATGATTGGAAATGTTGAAATTACTATACAGTTAATTTCTTGTTGTGTTGCTGTTTGGCTATAGGCATAATTGTTTCGTTTTTAGATTAATACGAAATTTTCTTTATTCTAAAGGACTTAAGATGAACAAAATCTATAATGTTTAACTAAAATCATTGTTTCTTGGGTAGCTTTTAGAATATATTAATTCTATTTTGTTTCATAAATATTAGGAAATATGGAATAGGTTGCCCCCAAAATGTGAAGTATGGACTTCTTGCCTCAGATAAAATGTCCCACCTTTGACATTTTTTATCTAAATGTAAATCATAGGTGATGTTTTCTTTTTTCTATCTCAATAG|GGAGATTATTTCTTTGGCAAATGTTTTGACGCAATGGAGGTAGATGCTTTAAATTCCTCACACCCTGTGTCTACACCTGTGGAAAATCCTGCTCAGATCCGGGAGATGTTTGATGATGTTTCTTATGATAAG|GTAAAAGTAGATTGAGTATAAGGATACAGTTTAGATACTAAAGTTATACATACTGGGGTGGAGAAGTTATAGGCAAGGTTGTGGGGTTAAACCCAGATTGAATGCCTTCTCTCTTGACACGTGCTGGCTGGAGTCAACTCTTTTAGGACTAACTTGCAGTTTGGCTCATAACCCTAAAGATTATTTTATGGGAAATTCTTCATATATTCTTTCTTGGGTTGAAAATTCATGGCTTACAGAAACTCTGCTTTTATTCATCATTCAACAAATAGGTATTGATTGAGCATCAACTTTGTACTAGACAAAAATTCCTGCCCTCGTGTAGCTTACTTTTCAAGGCCTTCAGCAATGGTTAATATTGTTGAGACGCAAATAATTGTCTTGCACAGTGTGCTTAGTAACAGAGTTGGGAATTGTTGACAACAGTTTAATGATTGGGAGATTTTATGTAAAATCCAGATTTCTAGCTTCTCTTGGGGAAGAAAAAGGAGGATTTGGCCACTGTATGTTCTTTGCAGCATAAGCTGGAGCTAAGTTGTTGTTCTCTCTTTGTAAGATCAAGGCTCTGCTTTTCCACTTTCCCTGCCATTCCCAACTGTTCTATTGTCTTCTCACCGAGGCTGAGAGTGTGTTGCCAGTTACCATTGTGCTTGGCTGTTGTTTTACCGGTAGCAAACAGAAAAGTCTTTCTTGCTTGCATGTTTCCATTGAAAGTCCAGGGGAAAAAGAATGTAAAAGAGCATTCTTCTTATCCTTGGACTACTTCCCTTATTTATATGCCCTGTCATGTGCCACTGGAGGCATTTGAGTTTGTGACTCACCATCCGTGGTAATGGGAGTGGAGGGGAAAAGAGCCCTTTACCAAGGAATACAGGGTGTCTGGGAAGACTCTTGTTCCCTTTCTCATTGTGACTCCAGCTCCATCAGCCCTCCATGCTCAAGGCTGCCTGGGCTCCCTGGACATATCCACTTTTCCTTCCCTGGCATCTACCTCTGCCTCCATCTCTAGTGCTCCACCCCTTGTTGTACTGGCCTCTCCTTAGTCCTGCCCTGGAATGGCAGTGGGAGAGCCAGGTAGTAGCTCAAGGTCCAATGTTTAATCTGCACCATTATCCCCACTCACATGTGAACAAAGGGAGTTGGCAGATGATGCTAATTTGCCCCATCGGGAGGTCTGGCTACTGATAGAAAATAAGGGCCTCAGTGGGCTCAGAGCATAAGCAATCACATTAGACAAATCTCCTGCCTAAACAGGTCCAGGTTTAACCTGCTTACTCTGTTTCACAAATTGCCAGACATTAACAGTGTTCCTGCAGTTGCGTTTTCAAAGAAATGTGTTTTATTGCAAAAGAATATGTGATTTCAGATGAGACTGCAATGAAACTATAGATAACAATTATTTCTATTATCTTTTCAG|GGAGCTTGTATTCTGAATATGCTAAGGGAGTATCTTAGTGCTGACGCATTTAAAAGTGGTATTGTACAGTATCTCCAGAAGCATAGCTATAAAAATACAAAAAACGAGGACCTGTGGGATAGTATGGCAAGT|GTGAGTATGTTTTTGAATATCTCTGCATTTGGGATTGACAGGCTTATCATCTTGTTTTGTTTTCCCTGCATTATGTTAATCCCTCTGAGGAGAATCATTGTTTTCTATAGAAATAAGAGTGATGTGTTTATTTTTGGTTTTTAG|ATTTGCCCTACAGATGGTGTAAAAGGGATGGATGGCTTTTGCTCTAGAAGTCAACATTCATCTTCATCCTCA|GTAAGTTTCTATATCTGTACATGTTCCCCCAAGCACATTCTTTTACTGCATATTCTTTGAAAGGCAGCTCTGTGCCAAACTTTCTGAGGTCCTTGATTATATCACCCTCATTCCAGATAAGACTGCATTTAAACTATTCCATACTCATAATCTTTTTCAATTTTTCTTAAAGTGTATCTATACTGGAGGGTTGCAGAGCTTTCCTTGGTAATGCTTCTCACTGATACTAATTTCTCTAGCTTCCCTTTTAAAGCAGTGGATTTATGACATGTTTCTATAGCAGATTACAGCTGCATTGTAGCAGTCAAAAGGATATGTCAGTCATTTACAGAGCTCTGCATTTGTACAAAGACAATGGCACTGAGCATTCTTGAATACTTGTCATGTGTCAGGCACATGTTAAGCACTTATATGTATTATCTCCTTTACTCTTCATAATAAACCTGTGAGCTGGGTACTATTACTATCCCTATTTTTAAAGTTGAGAACATGAAATACAAAGACATCTGATTGGTAAATTGCAGAGCAGAGATCTGAATTTAGATCTAACTCATGTTTTTAACTGCTAAGCTATAAAGTATTCATAACATCAAGTCACAAAACAGCCTGAGTCTCTGTTCATCTGGACTTGTGGGATGTTTTCAGAGGAAGCTGAGGGTGAGTCTGGAGATACAACAGAATTTTTTGTTTCTTTTTTCTTTTTATATGTTTACTTTCTTGGCTTTTTTGTTCCTAGAGACTGCTTTATTCAGTAGTTTCTAGATTTGTCCTGTGGCACATTTTATCTTTTTACTAATTCTTTTTTTTTTGAGACAGAGTCTCACTCTGTTGCCCAGGCTGGAGTGCAGTGGCGTGATCTCGGCTCACTGTAGCCTCTGCCTCCCCGGTTCCAGCGATTCTCTTGCCTCAGCCTCCCAGGTAGCCGGGATTACAGGCACACGCCACTACACCCGGTTAATTTTTGTATTCTTAGTAGAGATGAGGTTTCACCATGTTCGCCAGGCTGGTCTTGAACTCCTGACCTCAGGTGATCCACCCAACTCAGACTCCGAAAGTGCTAGGATTACAGGCATGAGCCACCGCGCCCAGCCTTTACTACTTCTTAAATCATTTCTTAAAAGCTTTTTTAGTATCAAAAAACAGCTCCTTTTGAGTTCCCACTATTTGTTGAGTGTGGGTCATCTCTGTGTTCTCACTTTAGAAACACAGGCTCTTGACTGAACATTGTTTCCACCTTGCTTGTCCAAAACCAGCATAGTTAGGTAGGTATTGAAAACCTGGCACTTTCTCTCCCTTCTCTTCCTTCATTCATTCACATGCCTGCTTTGTGCCCAGTGTTATTGCCAGCCCCAAAGTGTGCCCGGCAGAACTAGGTATGCTACCTGTCCTCCAGACACGTATGTGTAGTGGAGGAAAATGACAAGCAAACAGCTGGTGCTCTAATGGAGGTGTGATGGCTGGGGACATCACACTGAGAAAGGGATCAGAAAGTGCTCCTAAGAGGGGAGGTTGGCAACTGGTAGCGGCTGTTTGAGCATAAGCTGGTGTTCTCTCTTTTCTTAGAGTTGGGTTAAATGGGTGATGTGTCTGCCTTTTTGTGTACACACCAG|CATTGGCATCAGGAAGGGGTGGATGTGAAAACCATGATGAACACTTGGACACTGCAGAAGGGTTTTCCCCTAATAACCATCACAGTGAGGGGGAGGAATGTACACATGAAGCAAGAGCACTACATGAAGGGCTCTGACGGCGCCCCGGACACTGG|GTAATGCTCCTAGAGTAAAATTTGTTTTGTTGTCTAGGTAACATCTGCCTTGTAGGATGGAACCTTGCTTTTGAAATAATGCCCTTACCACTATTGCTAAAATATTTCAGCTGCATCTGTGTATCCTATGAAGTTGACTTATACTCCCTGCCCCCATCTTCCCAGTAGGATTAAGGAGGCTTTAAACCTTGGTTATTCTCAGTAAAGGTGACGATGTAATTACTTTAACATTCTCATATTTTGTAATTTGATATGATGGTAATTTCTGGTTACTGGCTTGAAATCAACTCCAACCTAAGCAACTGCTACTAGATTACAATAGTGCCTAGCATTTGGTTGGAGCTGAGGACAAAAGAATTTAGGTGATTTCCGAGAATGATGAGAGACTCAGTTGTCTTCTTCTGAGTTAGATTTGGAACCTGTTTGTCAGCTTAATGCTATAGAAGAATATTATTAGAAACAAGATGCTGCAACTTGATTGACCCTGGATGGATACCGTTAAAAAATTCTTTTATCTTGAAACAATTTCAGACTTATAGAGAAGTTACAAGAATAACACAACAAATTCCTATATATCCTTTACCTAGATACATGTTAACTCTTTACTCCTTTACTTTCTTCACCTCCCTCTCTCTCTTGCTCTCTCTCTCTCTCTATATATATATGTATATGTTTGTGTGGATATATGTGTGTGTGTATATATACATATATATGTGTATATATATACACACACATATATATATATAATATATACACCTCCCTCCCCCATCTTCCCAAAAGGATTAAGGAGGTTTTAAATCTTGATTAATCGATATCCATCTATCTACATACATACCTATATACATATGTCTGTATGTATATACTTGGCATATGTTTTTCTGAACTGTTTGACAGTAAATTGGTGACATGATTCCCCTTCACACCTAAATGTTTCAGCATAGTATTTTCTACAAAACAACATTTATGGCCGGTCACAGTGGCTCATGCCTTTAATACCAGGACTTTGGGAAGCCGAGGCAGGAGGATCACCTGAGGTTAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACACCATCTCTACTAAAAATACAAAAATTAGCCGGGCATGGTGGCGGGCGCCTGTAGTCCCAGCTACTCAGGAGGGTGAGGCAGGAGAATCGCTTGTACCCAGGAGGTGGAGGTTGCAGTGAGCCGAGATCGCACCACTGCACTCCAGCCTGGACAACAAGAGCGAAACTCCATCTCAAAACAAACAAACAAAAAACAAGAACATTCAGTTATATAACCATAGCCACAGTACAATTATCAAAATTGGGAAATAAACTTTATATAATACTATTATCTATAAATCCCATTCAACTTTTTCCAATTGTCCTAAAAATATCCTTATAGCAAAATAGAAAATAAATTCTGACTTAGGATTTAATACAGGATCACTTATTTTATTTAGTTGTTGCATCATTTTGGTCTCCATTTTTTTCTGAAACAACTCTTCAGTCTTTTTTGGGCTTCTCGATATTGACATTTTTTGAAGAGTACTGGCCAGTTATTTTATAGAATGCCTTTCAGTTTGGATTTTCTGCTGTTTCTCCATGATCAGGTTCAGCTTACATATCATTGACAGGAATCCCAAGGAGGTGATATTGTGTCCTTCCTAATATACTATTCAGGAGGCACATGAGTCCTGAAGTTAGGCCTCTGAGATGCAATCAGCCAATCACTGTAAAAAGGCTTTCAGGGTTCACGTGAGTCCAAATTCTTAGCTCTACTCAAGGAGTCCAAGAAAAGGAAATAGTGCCCGAGCTGACACTCCCTACAAAGCTGGAAAACAGACTGGGGATGTTTTGAGAGCTTGGCTGTAGCAGTTTTGGTACTATCTTGTTCTAAACTAACCAGTGCCTGGCACATGATAGGTGATTTAATAACTGCTTGTTGAATTGAATCAACAAATGAACAATCCACCTCTCCTTTATTTTAG|GTACCTGTGGCATGTTCCATTGACATTCATCACCAGCAAATCCGACATGGTCCATCGATTTTTGCTAAAAACAAAAACAG|GTAATTTATTTTGGAAACTACTAGTTAATTCAAGGAGGAATGTGAAAATGTGTAGGATTAAAGTGCTGTTTCTTTTGCTATTTATCTTGGTTAAAGCAAAAGGGATCAGATTGAAAATGCTTTCTGGTTTCAAAGAAGAATGGCCAATATCTTGCCAGATAGAGATTATAGCTGAATTATATAATGCTAAAAAGGAGGGATTACCAGTGAACATTCACTATTACTTTGTTAACATTACTGGGTTATATCATATGTATGGCCAGAGATGTAGTTGCCATTCTTTTCATAGATCTGTAAATCTCACAAAATTGATGTCTAAAGCTAGTGTTAGCTTTTGTCTATTGAGTTGCTGTTTTCTGGTTTCTGAGAGAAATAAGTGATGTTTTCAAATTTTTCAGTATTAAACTCTTCTCTAATTTCCTTGCCCAATCTTGGATGAGAAAAGCTTTCTTCAGGCAGGGGAGCAAATGGAACTTTGATTTATTTTATTTACTCTAG|ATGTGCTCATCCTCCCAGAAGAGGTGGAATGGATCAAATTTAATGTGGGCATGAATGGCTATTACATTGTGCATTACGAGGATGATGGATGGGACTCTTTGACTGGCCTTTTAAAAGGAACACACACAGCAGTCAGCAGTAATGATCGGGCGAGTCTCATTAACAATGCATTTCAGCTCGTCAG|GTAATACACGCTGCACAAAGTCGCGGTTTATTTCTGAAAGCAGCTGTTATTGTTCAAATTCTTGATTTCTAAAGACAAAAATGATTGATTGATAACAAGAAGTTGAAAGGTGTTTTCCTTTAAGTTCTTCTAACAACCCAAGATTGCTTTTAGTCTTATAATTAACCTCCTGTCTGTGACTACTAAGGCACTTGAAGGAAGGAATCTGTATCTTAATCTTTCTCATTAATAATTCAGATCATACTGAGGCGGAAGTGTAGATAACCAGGAGATCAACATCCTGGCGAAACTCCTTGCCTTGTCCCCGCGGCATGGTCCCGCAGCTTCTTCTGGCCCCGCTATACCGGATCAGGTTTCTCCCACCACTGGGTCTTTATAACAGCCTTGAAGTCTTCTTAGAAGACTAAGTGACACAGGCCCAAGTTGTTTGCATTACATTTCTTGGGTTAAAGAAGGGATTTTTTTTTCTTTCCTAAACCAGATACCTGAACTTTGCAGCATCTTTATGGAATATAGCTCATAAAATGTAGCCCAACCAAATGATCTCTGAGTGTGTGAAGACAGAATTAACACTACCCTTTTTTTTTCCTCCCCCATGACAATGGTTTTTAAGGAAATGCTCCCAAAGCCTAAAACTCAAACTTCTTCAACATGTAGGCAGACCTAAAGTCCTAAGAGGAAGCATGCAGGTGGGAAGGATTATCTCTTTCTCCTCATACCCGTATCCTTCTGGCCTAAATTTTGAGTGCCTTCTGGTTCTTCTCACCACCATACTGGGCCTCCTGAAGTGAGAAAAAGCAATGGGGGAGAAAGTTATGGGTTGCTTACCTAGCTGTTTCTTTCTGCCTAAAAAATTCCCCTCTGTAGAATATTTCTGCTTGAGCCTTAGAGGGCTTCTTCTTTTTTATTTTTTAACTTATATAATATATACCCTACCACTGAGGTTTTTAACACTAACCCTCTGTAATGAAGGGTTTCTAAGGATATGCAGTGTCTTTGAAATGGAGAAGAAAATGTGTTTCCTCCAGGAACTCATGGTTTTGACCAGATGATAGTCTGTTTGACCAGGCTTCCTAAAGGGTCTTTGATGGAAATCCCTTGCTTACTCTAGAAGTTCACCACTCAATCCAGTGTTTCACAGGGATAGATGCTGATGTGCTTCCTTTCCGGCTTATCCACGTTAGCCCTAGACGTTACTGTTTGAGAAACTTCCTCCCTAGTGTAAACATGCCGGATGTTGTGTTATATGTGATAGAAAAGAGACTTATATAGTAGTTATCTATCTGTTTGTTTATTTATTTTTTGGAGATGGAGTCTCACTCTGTCGCCCAGGCTGGAGTGCAGTGGTGCCATCTCAGCTCACCGCCTCTGCCTCCCGGGTTCAAGTGATTCTCCGGCCTCAGCCTCCCACATACCAAGTACCTGCCACCATGCCTGCCTATTTTATGTATTTTTAGTAGAGACGGGGTTTCACCATGTTGGCCAGGCTGGTCTTGAACTCCCGACCTCAGGTGATCCACCTGCCTTGGCCTCCTAAAGTGCTGGGATTACAGGCATGAGCCACCGCGCCCGGCCGTAGTTATCTATTTATAAGGGACATTTATATCTGATTGCTTGTTTTGCTAGAAGATTAAAATATATTTTGAAGCAAAAAGTTGACTTCCCGGATTTACTTTTAAACCACTAACCACAGTGTCTCTTGGTCAG|CATTGGGAAGCTGTCCATTGAAAAGGCCTTGGATTTATCCCTGTACTTGAAACATGAAACTGAAATTATGCCCGTGTTTCAAGGTTTGAATGAGCTGATTCCTATGTATAAGTTAATGGAGAAAAGAGATATGAATGAAGTGGAAACTCAATTCAAG|GTAAAAGCCTGAAATAAAAGTTATGTAATTATTATTTGTGTTAAAAAGTGTTAATCATTGTGTGTGTATGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTATCTTTATATACATATTAAGGAAAAGCAAATAAATATTAGTTCAGTATTTAGATGGACTAGCAAGATTTTGGTTTTATTTGGCACTAAAAATAGGAATGGCATTCTTAGACTCTATTTAATTTGCATGGATTTGTTATTTCCTTCCTTTCACAAACCTTTTATTTCATTTTCCTTTCCTATTACAATGGCTGACGTTCAGATTTTCTGGGACTCTTCATGGTACTTGAGGAAGAGGCACAAAATTGTTATGCTTGGCAAAATGTCCTGAAGTCTTGTTGCATAATTTGCTCTCAAAATTTAGTCATGTAAACTCTCTTTGAGTTTGTTTTAAAAGTCGCAATTCATCCTGACTTTAAGTAGATGGATATCTTATAAAAGTTTGTTATGAAAATATGAAAGTCATTCATCATTATTTATTGTTTTTGGCCCTTTTATAATTAAACAGTGCTGCATGATTTATACAGTAAAAGTTAGATTATGCTTTAAAAATTAGCCCCCATCATCTGAGACCATAATGGATTATATTAACATGAAATGACCTGTGACAATACTGGTCCCTGTTTCCCTGTACAACGCCCTCAG|GCCTTCCTCATCAGGCTGCTAAGGGACCTCATTGATAAGCAGACATGGACAGACGAGGGCTCAGTCTCAGAGCGAATGCTGCGGAGTCAACTACTACTCCTCGCCTGTGTGCACAACTATCAGCCGTGCGTACAGAGGGCAGAAGGCTATTTCAGAAAGTGGAAGGAATCCAATGGAAACTTGAG|GTCAGTCCTTACTAAATAACCAATTTGTTGATGTGAAGGGCATCTTTTCTGTTTTCCATCATTGGTACTAAACATTAGGGAAAACAAAAAGTTTAAGTGTCTCCCCTGCTGCCCTTTTCTGGAAAATAAATTGCTTTTTAAGATTTATCTATGTATCTCGTAACTTTAAGAAATGCTAGGAGGGAACTTCTATGCATAAAAGTCAAATATCTGGGGAGTTAGGATGGTATGGGAATAATTCTCATTTTGTATAGGCAATGCAAAATCTTATTAAACGGTGACAGCCATGCCATAGAGAAACACATGCATTGTATTTTATAGTAGTTTCTTTGGGCTTCCAAATTCCCTGACAGGGATGCATAGTAATTGCTGACACTGTCTAAGCAAGGTAGGATTTTACCGTATCTGGAAATCCCTATTCTTGCTACACTTCAGTCCTTTTTACATTTGGAGCTTAAACCCCACCCAGGAAACATTTATGCCAACAACCTACATGCTTCAGCTTAAAGAAATCCAAACAGTGGTCCTACTGCCGTCCTGACTTGTGATCATGGTGTAAATTTTGAATATAGTTTGAATTTCTTTTGGCTTGAAAGGTGATCCCCTATGGAGTCACCCAGCATCATTAAAGTATTTAAATATGTAGGTATTTATAAAAATGGCATTTCACATTTTTGAGAAGGCTACTTAAAACCTAATTTTAGATATTTTTCTCTTGCCTTTTTTTTATAAAGTGTAACTAGGCCAGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCTGAGGTGGGTGGATCACGAGGTCAGGGGATCAAGACCATCCTGGCTAAGACGGTGAAACCCCGTCTCTACTAAAATACAAAAAATGAGCCGGGCATGGTGGCGGGCGCCTGTAGTCCCAGCTACTCGGGAGGCTGAGGTAGGAGAATGGCGTGAACCCGGGAGGGCGGAGCTTGCAGGAACCGATATCGAGCCACGGCACTCCAGCCTGGGTGACAGAGCAAGACTCCGTCCTAAAAAAAAAAAAAAAAAAAATTGTAACTGAAGGAAATCAGTAGTTTTTGTTTATATTTCCAGGCATGCTTCTCTGATCATGGGGAAAGTTCAGCAGTTCATAGTTAGTCACCATTATGCCTCTAAACATACTTACTGCACCTCAGATTCTCACCGAGTGTGTTCCTTTCTATAG|CCTGCCTGTCGACGTGACCTTGGCAGTGTTTGCTGTGGGGGCCCAGAGCACAGAAGGCTGGGATTTTCTTTATAGTAAATATCAGTTTTCTTTGTCCAGTACTGAGAAAAGCCAAATTGAATTTGCCCTCTGCAGAACCCAAAATAAGGAAAAGCTTCAATG|GTGAGTCCGTCATTCATTCATGTTCATGTGGCCAAGGGAAATTAGATTAGATTAGATCATTCTGGCATCTATTTTTGTTTTCTTGGCCAGGAATTGCCTATCCTGCTGGGAACATACTGCAAGTCAGCACACATCTAATGAGAAAGGCAAATAATTAGGGAGAACCAGGTGTTAGGAAAAATATTTAAAGGTGTATACATTTGTGCATCTTTAAAATGACAACATATTAAAAATATTTACTAGAGCTACTCTTCCTTAACTATTCTTTTGAGCAAATGAGAGGGTATATGAGAATGGGTGTTATGAACACTATATAAATGTTAGTTTTTATTATGTAGCTATAGTGGGTTTATCAGGATTGTGGTTACCCATTTATCTTTTTAACCTTTGTTACTGGCAGACTTTTTTTTCTTTTACAAGTAATTTTATTGCAGTGATTCATAACTGTTTCATCACCTATTCATTGCTAACATTTCTTTTCTTAATCCTTCTAG|GCTACTAGATGAAAGCTTTAAGGGAGATAAAATAAAAACTCAGGAGTTTCCACAAATTCTTACACTCATTGGCAGGAACCCAGTAGGATACCCACTGGCCTGGCAATTTCTGAGGAAAAACTGGAACAAACTTGTACAAAA|GTAAGTGGTGCCAAAAATTGTGCTGTGACTGGATAAGTTCATAACCTTACTGTGTTTTAGCCTTGCTGTTTGTAAAAGAACAGTAACAGTCTAAAGGTACTTTTTGATTGAAGATAGGCAGTAGAAATACCTAAAATATTTGTAGAAAACATAAAACTGGACTTCAGTGCTAACTAGTGAATCTGGACAGGGATGTTTTCCATTCCATCTGGCATAACCCCTTCCTGAGCCCATGGACATATCTGAAGCCTTCCTCCTCACAGTTCAGCCCAGGCCTTCCATGAACACATTTGCTTGTTCACATCTGTCTTTGTCTAACTCTTATAGCATTTCCTGCTTCTGTCATTTTCTGTTGGATACTTAACCTTTTATTAGGCTGTTGGTGTGTATTATTCTTTACAGCTAGATCTTAACCCATTGGATAGACATCATATTTTGTATTTTTCACACCGATCAGTTTTTAGCTGAAAGCTATTATATATAGGAGGCCCTTAAAATATATGTTAAATGAATAAGTATTTCACAACCCGTTTTTGAATATTTCCCTCTCTAG|GTTTGAACTTGGCTCATCTTCCATAGCCCACATGGTAATGGGTACAACAAATCAATTCTCCACAAGAACACGGCTTGAAGAG|GTAAAAAAAAAAAATCTATATATATATTTTTAAACATAAATGAAAATTAGCTAATTAATATGGGGTAGACAAAATACTTTGAGGGTGTGGTGAGTTAGAAATGGATTGTCATTTAGAAGTTATTTTTGGATGCTATGGTGTTGACAGCAGCATAAATCAGTTGCAATTAAACTAGTGAAAACTGTGCCTCTTCTCAGACACGTTAAGAGGTCTTAGCTCTGCCAGTAAAAACCTTAGGACTTGAAGAAAATTACTTGAGACAATTAGTCCTTGTTTAAGGATGTTAAAAGTGGGCACTGAGGTATAAATGACTGAAGTGGTCATCCAGCTGTTGCTAAAGAGGCAGGTCCGGAATCTAGGTTCTTTATAAACCATCCAAGTTCATGGTTCCCTGTCACTTATTATGTAACCATCTACCTTCAGAAAGAATTTCAGGCAGCTTTAAGATCCATGTATAAAAGAAATGTTAAAATGAAGGACAAAAAGATACATAATGGACAGGTGATAACCATGTGAGGGACTTATGGCTGAGGATAATTCTCGCAATTGTCCCCTGAATTTATCACAGACATTCCTGGAAGTCAAGGAAAAATGAGGAATCAGTATGAGTTATTCTCATTGTCTGGTAAAAGAGAACATGAAGCATACACATTTTCTACAAAAGCAGACTTTTCTTCAGTCCTAAACTCGAGGACTTTGATGTGTGGGCCATTGAGTTCTGTGGTGTCCCTTCTATAATAAGGTTTCCTATAGTAAAGTTGTGGAACAAATTTTACAAGGCTCTTAAACTGTGGTACATTTAGATAAAAGCTGAGAGGCTAATATTAATTTTTTTAGGAGTGGTTCCATTGTGTAGTCTGGTTACACTATTTTCTAGCTTGAGATGGGGAGAGAGCTTTGGAAATGAAAAAGAATGAAGGGTTGTTACAGTCTCGTAGCTTTTTTCCAGTTTTCAGTAGCCTCATCCAGGCTTGTAAAATTAACTTGCATATAATAATTCAAGCTTGAACTCCAGCGAGGTCTAGAGCAGAGATACTCTGTCTGATGACTGAAGAGATGTCCAAAGCCTTGACCAGAAAGGTAGTCATCTATAGACAAGATCTATAGCAACAAAGCATTTCATTCATTCACTTGTATTCTCATTAATTCATTCAAGACATATATGCCACACATGGTTTTAGGTTATGGGGCTGTAACAGTGAACAGACAGACAAAACCTTTGCTGTCACGGAGCTGAAATTTCTGCTGTGGGAATCAGACAATTACAGACAAATCAACCAATGTCAAGTAGTTACATGCTCTGAAGATGAATAAGGCAAGGAGAGTGATTGGGATAGAGGGTTGGTGCTGTTTTATAGGATGATTAGGAGACATTTGAGAAGATATCGAAGGAAATGCCAACATCTAGAGGAGTATTGTGTGCAGAGGAAATATCAGGTGCGAAGATCCTGAAGCAGGGCATGAAAAGTAAAGAGCCCGAGATGTTATCATAATGGAGAGCAAACTACGGACATATAAAATACGGGAGTCATCCTGAATCCTTTCTTCCCCAGACATCTAACTACGCAAGTCCTGCCGGTCTTCCCTCAGCACATGCACTGGATCTTGCGCACTTCTCTCTGCTCTTACCCCCATTCAATCCTTTCTCATCTCTTGCCTTGCCTATTGCAATAGTGTCCTAACTGGCCTCCCCCTTGCCACTTTGCCAATAGCAAAACATTTTCCACATTGAAGACTGCAAGTCAGCCCATAACTGACCCTGGCTTCTTGAAACCCTGCCTATTGTTCTTGCAGTAAATTCCAAATCCCTCCTTCCCATGTGTCCATGACCTTGAAGACCTGGCTTCTGAGGACCTCTCTGGTGCCAATTCCCACAGGCGCTCCTGGCTTACTCTCTCATAACACTGTAGCCACTCAGCCTCCTTCTGCTCCTGGGCTACCACATGTTTCTCCTGTCTTTAAGCCTCTGCACTTTTTATTCTCTCTGCCTGGCTTGCCCATTCCCCAACTTTTCACTAGAAAGGGGCATTTTATCTTTGAGAGGTCTCGGCTCAGATCACCTTCCCCAAAACCATGCTGTTCAATGCAGGCCACATCTCCTCAGTTACTAGCATACCCCTGGCATATTTTCTTCATAGCAGTTTATTTACAGTAATCATATTTACTGCTTGTATACCCCAAAAGACTGTAAGCTCCATGAGGGGAGGATGCTTGCCTGATTCACTGCTGCATCCTCAGTGTCTAGCACAGTAATTACAGCAGAGGGATGAATAAATATCAGAGTAAATATTGGTGAACAAATGGATGAGGTTAAGTCCCTGATACATGCTGGGTTTGGGGCTTTGCACTTTACCTGCATTATATAACTACATCCTTCACCACTGCTCTTCATTTTACAAATGAGGAGACTGGCTTCCATAGAAGTTAAATCATTTGCTCAAATCAAGTTAGTAGGATAAACCTGTTTTCCTATCCCTGTATTGTGCTACTTATTGCACAGAAATTGTTCTTAAAGAGCCAAGTCTGAATCAAATATTCAGTGGAGATGTTGCATTTCCAAGGCAGGTGAAGACAGAAGAGATGATTTTGGGTCAGGACAAGGGTAAGAGTAATGGTTAACAGCTGGCTTCGCTGTTGTAGAAGGTCTGTTTTTAGTGGAGCTATCATGATGAGCTCCTTTAGGATTAATGATTTGAACAATTAGTAAAATTATTGATGTGCTTTGTTGAGTAGTGCTCAGTGATTTTTAATCGACTGTGAATTAATCTTGCATTCTGAGAGCGTATGAAACAGTAGGTTGAGCTACAACTTAAAGTAGAGAGTGAAGTGTTTTCCAGGTTGCTACAGCAGGAGGCCATGCCTTCTGCTCATTGTAAAGTTGCACTGTATTGACATATAATTTTAAAACAACTCTGCATCATTTAAAATTGAATTCTGATCTTTCTAAAACCCATTCCTACTCCCCTCTTTATTCCCAAACTAATAATATGGTATTATGACATGGATTTCTAAGGAACTGGCTGTCTGGAATCTATGCTGAATAAATAATACATCATGGTCTACATTCGCTTCCTGAATACCAAATCAAAAATTGATGGATTAATGCTGTGAAAATTTATGGGAAAAGGATAATAACCCTTTAAGGTGAAACAGAATCGCACAATGGTCAATCTTGTGGCAAAGCCAGCCTATCTGTCATGTGAGCTCAAGGGGTGAATTAAGATACCCACAAATGAAACAAGAACAAACATATTTTTCAGGAGGTAGCCAAGAGTTTCTTATCTCTTTAAAATCATGGGGAACTTTATGTGGATTTTATCTTGAGACAACAATACATGAATTGTAGATTAGGATAAAAAAAATGGCAAGGTTTGGGTCTTACCGCCATAACTTGCTAAGAATCCCATTCCCCACAGTGTTTCTATAATAACATACAAGGCCAGGCCCATGGAGAAAGACAGCACTTACGTGGTGGAAACTGTTTTGCTTGGCAAAGAAAAGACTCTGCACATTCTGCTTTTTAGATATCATGTTTTTAAAAAACGAAGTTTGCATCTGTGAGAACAGAACAGAATAACACATTAAATAGACACAATTAAACCTTAATTATAAAATGGTAAACAGTGAAGTCTTTTAAAGTCTGGCAACTGAGAATAAACAAAAGAACTGCAGACAAAAAACAGAACTTGACATCATGAGGCATGAGCTCATTTCATACAGCTTATGTGTACATAATCCTATTCAGACAGCTGGGACTGCCTTCTATATAGAATTTTGACAAATGCTGGAATTTTGGCTTCAGTTTTAACTAAAGTTACATCTGATTAATGTGATAAAATTAATTTTTTAAAACCCACTTTTTCCTCACAAG|GTAAAAGGATTCTTCAGCTCTTTGAAAGAAAATGGTTCTCAGCTCCGTTGTGTCCAACAGACAATTGAAACCATTGAAGAAAACATCGGTTGGATGGATAAGAATTTTGATAAAATCAGAGTGTGGCTGCAAAGTGAAAAGCTTGAAC|GTATGTAAAAATTCCTCCCTTGCCAGGTTCCTGTTATCTCTAATCACCAACATTTTGTTGAGTGTATTTTCAAACTAGAGATGGCTGTTTTGGCTCCAACTGGAGATACTTTTTTCCCTTCAACTCATTTTTTGACTATCCCTGTGAAAAGAATAGCTGTTAGTTTTTCATGAATGGGCTATCGCTACCATGTGTTTTGTTCATCACAGGTGTTGCCCTGCAACGTAAACCCAAGTGTTGGGTTCCCTGCCACAGAAGAATAAAGTACCTTATTCTTCTCATTTTATAGTTTATGCTTAAGCACCCGTGTCCAAAACCCTGTACCCCATGTTTATCATTCATAAACTGTTTCATCAGTCTCCTCGAAAGACTCTGAATAGTCGACTACTGAACAATGAACACCTGGATCTGAGACTAAGCCGGACGATGACTGGGTTAAAGCTCTCCCGGCTCACCCCTCCAGACCCGCTGCCCATCCCTCTTCCTTGCTCCATGCCCAGGGGCTGACTTGTAAAGGCCAAGTCATCAAGCTTTCTTGCCCTTTGGATGTTGGTCAGTGGGGAGCCGGAGAGCTGGAGCTGGGGTCGGAGGAGGTAGTAGGTGGAGGTGTTCTTCCCTGATTCCCTTGCGGGATGCCTCGGGCTGGCCTCCCCTGAGGGTCTTAGCTCCGAGAGGGGACCCTCTTTTCCACACAGCCTTCTCCACCTCTGGATTTTGGTAACTGCTCCCTCCTCATCCCTTCAGGATTAGTGGCCTCAGTGGGAGTCTGGCTTTTACTAGTCCTGGCGGACTTGTGGTTTCTACATAATGTGCTCGCACTTTTGCAAAAAATCTTTTTATAGAACCCTCCTCAGATAATTCTGAGTGAGTGTCATCTATTTCCCTGACTGGTACAGTATCTCTTCTGAAAAAGCAGAGTGCATTCAAGTCTGTAGGAAAACCCTTTTCTTAGGGAGGTGATTTTTTTTCTCTCTCTGCTTCTTATTTGGCCTACTTTACAATTTCTAACTAACTAGTTATTGGCATTTACTGACAGTAAATTATTGCAGTCACCAATAAATGATAGTACATTGTGAAACAAAATATTTGCTCATATTAGCAAATAGGACATTCTTTGGCTTTGAAGTCTTTCTTTCTTTTGTGAAGACTTCACACACGGTTGCTTCAGCACACAGTTGCTGCTCAGGTTTTATGTATAGATGATAATAATAGAAAGCACAGTTTACTAACATGGTAAACCAACGGAGTTCAAGTCAAGTCAGTTAATACCCTAAGAATTAGATTTTATTTCTTATTCTGAAAACTTGCTACACAGGGACTTATCTAACCCATAGTGTGCTCTGTTGCTGACTTGATTCAAGTTGCAGCGTGTTTTGCGCTGACTCTAAGGTGCGGAAATCCTCACACCTGGCAAAGGAGAATTCAAACTGAACTTTTTGAATATAAGGCAAAAACTTCAAGATAAGGGAATATGATTGATGATTGGTACGAAAAATGTCAAAATGTGTTCCCCTAATACACGACAAAATAGAGTGACTTCTGGACATAAATCTGCCATTTATTAAACCATTCACTACAACAAATAAATAGGTATAAAAGTGGAATTGGAATTTTTATACTTATTTGTTGTAGTGAATGGTTTAATAAAAATAGAAATCACTGGTAATTTCCACCCCAAACTAAACTATTTCCCTTCTTTTAAAAAAATACACAACCAAGATTTTAATGTAAAATATTTTGCTTTAATTGTATTTTATGCCTTGATTAATGAAACATGGAAATATTGATTTTCAGTTTTGGTCACCTGAGGAACCTATCTTTGTTTGCTTTTGGAAAAGCCCATTTTCTAAACAGATACAATATTGCCACAACAATGTGCAGAAACCTTTTTGATAATAAAAAATTGTTCTTTGCCTCTAAGTGGATATTTGCAATTATTTTCTCTCTCCTAACTAGACTGTAAAAAGGGCTGCTTTAGATCCTGTAGCTTACTCCAGTTATTAGTTATTAACAAACACCCAAGTCTCGAAGATATTTCTAATTAAAAAAGAAGGCATATTCAGAGTTCTTTTTAAATAAATGTTGTTTACTTTTATAGGCATCTTTAAACTTCTGGATTTTGGTATGCCATTTAAAAATACTTCCAGATACACATGGAAATTAGTAATACTGAAGCCGTATCCTTGCAAACACATCTGTCAGTGTCAAAGGTTTCAAGGTTTTTCTTAAAAAAAGAAAACAAAAAAGCAAACACCTATACTGCCCAAATGGGAGGATTAGATACATGGTTAGAAATCCCTCAGGAAAAGTGTTTTTTCTTTTCTTGTTGCTGCCTTAAAAATAGAATAATGACTATTTCTGATGGATAGAGACATAGCATTTTAAGCTGGTGGTGTGTAAAATCCCATAGGTATGTGCATGACTTTCAGAGAGTATTTGGGGGGGAGGTTAACAAGATGTGGTGCCATTTATAAGCAGTGTTATTGTTTTTGCTTGCCCCGCTGCCACAAGTCAGCTAAGTCATAACAAAAGCTTCAAACTGATGCTAAGGAAGGCCATGCCCTTTGGAAACAATAAATTCCCAATCTGTTTTATGTTATGTACCTGACATCTTTTCCTGCATTCTCTACCAGGAAATAAAGATGAAATTAAATATCAAAATTCTAATCGATGATATCAGTGCAATGTTCAGGAACTATTCATTAAGATATTAGAAAACCATTCAAAGTGGTAGGACATCAGAGCCTACTTCTTACATTGCTGTGGGAGAAATGCAGGTTTCAAATTTAATATATATATATAATTTTTTTAAAAGCAGAAGTTTCTTTTTATATTTGGTAAACTTAAGTTCCATAAAGCCAGACGCTATACAGTGCAAAGGCTAATGTGGCAATAGCTCTAAAGACACAGTTGCTGCTCAGGTTTTATGTATAGATAATAGAAAGCACAGTTTACTAACACAGTAAACCAACAGAGTTCAAGTCAGATGAGTACCCTAAGAATTAATTAGATTTTATTTCTTATGCTCGAAATTTGCTACACAAGGACTTATCTAACCTTTATTTTGCTCTGTTGCTGACTTGATTCAAGTCTCAGTGTGTTCTGTGCTGACTGTAAGATGCAGAAGTCCTCACACCTGGCAAAGAAGAGTTCAAACTGAAAAGGGGTTTGGTGCTTCCTGGTTTGTCCAGGTTACCTGTTATTAATTTATTATTAGCAGCCCAAGAGGAGATATGTGCCCAATGTACAATATCTTATGTTTGACTTATAAACATTATCCCAAAGCAACATCAAATACAGTTCAAAAGCCCAAGAGGAAAGGGGGTAATAAGAAATCAGAACACTGAAGAATGTTTAAAACATTGTTTTCTAAACACTAACAAAAAAAATTAAGGGCAAACTGAAAATACAAATGAGATTTACAGGCACTGTGTGTAGAATGTGCAAAAATTCACTTAGCTTTTCTTTTGTTTTTTTGGTGTTGCTTTAAGAAACTTTATCAAATATATTTCTTACAAATATAAAGCTTTCTCTCCCAATTGAAGGCAATTAAAAAAATTCAAAGTTTATCAATACTCAGTACACAGGTGAACCAGTCAAATTCATTTTCTTTCTGGAAAAGAATAACAAACCAATATTTAGGATGTTCAGAGACTCAACAAAAACCATTCTAGAAATCACCCAGAACAATTGTTTTCTGTTGCCAAAGCCTTTTGTTCTTCAAAAGTCACCATCCACCAGCTGAAGATTTTACATGCAGATACCTTAAAAATTTCAAATAAAAAATGCAGTGAATCATTTAATATGTAATTTTCTTGTTACAGACATAGTAAATATACCACTTAGCAAAAGCATTGTATAACAGACAGAAGGAATTTCCTATAAGTAAACATGAAAGTGGATTAATAGAATTTTTTTTAATTTTGGGAAAAATGTTAGAGCAGTTCTACCTAATATAGCTCCTTTTTTCCTAGGAAATAAACATGGATCATGGTGAGAGAGCTGAACCCGATTTAACCTATACTTTGATTTCTTTTAGGCTTTGGTCAGTAAGTGCTTGTATGCTTTTAAGGCTTACATTAAGCCCTCTCCTTTCTGAAGATTAAGATAAGGGCCCAGTTCTGAAGATCTCAGAAATCCCTTCAATCTGTCCAGTTTCTTCAGCAATTATAAATTAGAATCAAACACAGTACTTTACTTTCCAAAATAATGACAAATAAAAATGGCCAATCTTTTCCTTTGCCCTTGTTCCCAAAACCCTGTAATTTCCACCAGACTCTAAAGGGTACTTCTCCCCACCCCACCATGTATTGGTATTCTGGGGAATGCCAATGTCTTCAGGACATTTCGCTAAGTTGTAACATGCTAATTTTGCTTTGCCACAATAACCTGCAGTAATGAGCATTTGGATTTCAAAGATTCAACTAGCCTCAGATGGTCATTCTAAGTGCCTGGCCTAATATTTTTAAAAGCTTTTAATTATAAGATTAAGGTTTTTAAACATTTCAATAGCAAACTCATCCCATTTAGTGCTTTCAGAGAATGACCAATTACTCTGTAGATCTTGCAGTATGCATTTCATGCCAATACTGTAAAGTGAGCATGAATTACTCAAGAGGTGGACTTCACTTCTCTCATCTATAACACATAAATTGGCAGAAGATACAGTTGTCTTCATTTACAAATAAACACCCAACTTACCAGATACCTTAACTTGTATTTCTTTAGTCATCTTTTGGCTTGGAAGTTTCCTCTGTTGTCTAAAAGGGAAAGCAAAACCATCCGTGAGCTTTCTTTTCTGTATTAAGTATGAGGAGATGGCCTTCTCAGAATTAGGGGACAAAAGATGGCAGTCAGTGGGGAAAAATAAGATGGTCCTTTCAGTTTCTCTTCTTCATCTGGCCACAATATTGTGAGTTCTTCCTCCTCAAGCTTATAAGCTAAAAATAACCTTAAGTGATCCTGATCCTAAATGTATCACTCTCAGCTTTATTTTTTTAAGGCTAGGGTAGGTTATAAAATATGCATAGGTGTTCTTAAATGGCAATGTTTCATTCTGTGGTGATCCTTCCTTTCCTGTACATAGGGTCATGGCTGTCCAGTAAATCCACTCATCATAAAGGGTTATTATGCTTTCTATTATTTGTTAAAGGGCTGATTAAGTACTTCGTATAACTGAAATTAAATAATAGGCAAATAAGTATTTAAACTGAGATATAATGGCATAACTGCCAATTATCAATGCTGTTTCCTATATCCCTTAAGGGAATCTACTACAGGAAGATTTTAATATATTGTTTTAAAAGCTTTGGTGTAACTGGACTATTGTCTTTAAAGCTACACCTTTAACTCCTCCTTATAGCCAGGGGATCATAGTAATAATCATTTAAATCATATGTTCTTGGAATTGGAAGGGACCTAGAAGTTGTCTAATCCAATTTTCATTTCTCATCACAGCTTGAGTAATTTTAATAATAGGAAGTTTCCAACTTCCATGTTCTAAAATCTATATAAACCACATACTATGGTGGTATTTTAATTAGGGCAAGAAAGACAGGCAAAACACAGGCAAATTGGGTTGTGTCAAGAACATTCATTTAGGATTTTAAAGCACCAGTTACTTAAAAATATATGTATTTATACAAATTCAATTACTTTACCAAGCGATGCTATGGGATAGATAGTACTTTAAGAAATTTTATTCTAAATAGATTCCAGTAGGAAACTAAATGACTGAAATGATAAACCCTACTCTGTGTAACTGCTAAACTAATTTGTAGTATATTTACTGCTCCATTTACCTTTGCTGAATCCTTCGCTTTACCTCCATTCTTAGGTGCTTTGGAGCTGGAAGCAGCCTTCTTGCACTTATCCTAAAGCAAAGGAAATGTAATGAGGCTATGGCTATCACCAATCCATCTAGCAATTAACTAGGCTGCAGTTAAATTAATCCAGACCATTCTGAGATCTCCAATTTAATTAAATAATGGAAGACTTTGGGTGTTTTTTCATGATTTTTTTTGAACAACGGTGTCAGAGAATTGTTTAAAGCAGGAGGGAAAAAGGATAGAAATAAGGAGGGAAATGTGGGTGACATTGATGCTCTAATTCCCGTGGTGCCTGACTCACCCAGACCTTATTTTGCTAATCAAAACAGAGCTTGTCAATAGATAATAAATGTCGGCAAGGGTGTGGAGAAAAGGGAACCCTAGTACACTGTCAGTGGGGATGTAGATTGGTATGGGCTATGATGGAAAACAGTATATAGGTTCCTAAAGAAATTAAAAGTAGAACTACTGTATGACCCAGTAATCCCTCTTCTGGGTATATACCCAAAGGAGATGAAATTATCACCTTATAAAGATATCTGCACTCCCATATTCACTGCAACATTACTCACAATAGCCAAGATATGGAACAACCTAGTTGTTGATGAATGGATAAAGAAAATGTCATGTATATATACATAATGGAATATTATTCAGCCTTAAAAAACGATATCCTACTATTTGTCACAACATGGATGGACCTGGAAGACCTTATACTAGATGATATAAGCCAGACACAGAAAGAAAAGTGATTTCACTTATATGTAGAATATATATAAAAGAAAAAGCTCAAAAACACAGAGAATAAAACATGGCGACCATGGTAGGGAACAGGAGGAGGAAACAGAGATATAGGTCAAAGGATACAAAATAGCAGATATGCAGAATGAACAAGTGTAGAGAGTTAATGTATAACATGAGGACTAAGGTTAATAAAATTGTATTAGGGATTTTGTTAACTAAGTAGATTTTAGCTGCTTTTGTCACAAAAAGTAGTTGTGTGAGAATGATAGATATGTAAATCTGCTTCCCTACAGTAACCATTTTATTATTTCTATGCATCCCAAACTACCATGTTGTAAACCTCAAATATACACAATAAAATGTATTTAAAAAACAAAATAGAGCTTGTCTCGATCAGGACTGGCTTTTGTGTACCAAAAGGCAAAAAAAAAAAAACAAAAAAAAACCCTGTTTTCAGTGTTATGGGAGAGAAATGAACAATGGGAAACAACCGAGGAAAGCTGGAGCAGGTTACGTATAAAAATAAAGTCCATTCACCAAAAAAGGCATTACTTACGAGTTACCAGGGGTGAGAGATAGGATGCTGAAGTGGTCTAGAAATTAAGCTACCCAGTATGGAAGGGCTGACAATTCAGTGATCGAGAGCAGTGCCTTAGAACAGCCAAAACAATAGCAAACTGAGATCTGCAGAATTAACTCTCCTGAAAATAACAAGGAGGTACTCATTTCACGTTTCCTTCTATTTGATTTACAAGAGGGTGTAGCTTGAGGGAAAATGCCTCACACTTGTTGAATTACACAGTTGTTTCTCATTCACTTTTAATCACGTTTTGAGCACCTGCTAAGTACCAGGCATTTTGCTAATGAGGAGCACAGAGGTAAAAGACACATCACTACTGTATGAAATGCGTAGCTCAGTGGTGTGATACACAAGCACAGAGAGGTAACAGAGAGCAAGGAGGGCATGGAAGAGAGGCCTCTAACTTTGGACTGGGAAGGGAGAAGATGTAAGACAAGAAAGTCTTCCCTAAGGAGCTGATGCTTGAGCTGTGCCCTGAGGAATGAAGAGTAGACCAGTTGGGCTAAGCAGACAGAAAAGGGGAGGAAGCTCCAGAGAGCAAATGAGCATGAGAGTGCCTGATGTAGTTAGGGCCTGCTCTCACTTTAAATGAACACAGACATAGCATTATTGTGGCACAACCATATAGTGTGGAGATAAAAAATGGTGGCTATGGAAATTACAAAGTAGCAGTTAAGAAATAACGTTAAGCAGTGTTTTATAAGTGGACTGTAAGTATAATTATGTAAAATATACATATAGAAAAAAAAGGAAATCCACAAAATAATACTGTTTTGGGGGCAGTGGAATTATAGGCATTTTTTCTTTTCTTCATTTTCAGGTTCTCTATATCATCGTTTGATTCATTCTACAGTTTAAAAATTGTAAGGGCCAGGCGCAGTGGCTCAAGCCTGTAATCCCAGTACTTTGGGAGGTAGAGGTTGGCAGATCACTTGAGCCCAAGAGCTCGAGACCAGCCTGGCAACATGGCGAAACCCCCTCTCTACAAAAAAATACAAAAATTAGCTGGGTGTGGTGGTGCACGCCTGTAGTCCCAGCTACTCAGGAAGCTAAGGTAGGAGGATTGCTTGAGCCCAGGAAGCAGAGATCGCAGTAAGCTGAGATCACACCACTGCACTCCAGCCTGGGCTATAGAGTAAGACCCTGTTACAAGACAGACCGATAGATAGATCAATCAATAAATAAAACTTATATGTATGTACACATACACACACACACATTTCAAAGAGTGAAATGTGAAAAAGCACAGTACCTTTGCTGTGTTCTGTGAGGTTTCTGTAGTGGAGGGACAGCTGTCCAGATCTCCTGAGAGAGCATCAATGGGGTCTTGGTCATCTGCAGGTTTCTGAGATATGAGTAGAAATAACCATCAGTGAAGAAGCAGAAGGCAAAATGCAATATGGGGTCTTTTCCCACATCACTTACAAATAAAAGATGTTTCTATAAGAAAAAAACTGACTGACATTCTTTATTAATAATAATGTATTGTAAAGGAGATAGAAAAACAAGAAAATCTTGATGGCTTTTTTTCCATCTACTCTTTAATACACGTTGCTTAGCATTCTCTGAGCCTCAGTTTTCCACCCTAAAGGGCTTTTGTGAAGACTAAAAGAGAGGTAAACAATAGTCACACACATTTATATGCATGCTTTGACAAAGTACCAGGCACAGAGTAGGCATTCAATATGTTTTAGTTTTCTAAAATGCCAAATACCCCTATGGCTAGAATAAAACAAAATTTAATGGAAATATGTTCCTATGGTCTTTACCTTTGAATCCTCTGATTTCTTTGTAGGTGGCTTCACTGGTTTGTCCTTAAAAAGAAGGCAGACTATTGGTTAAGCATAGATATCTGTAAAGGTTTACTTAGGTTTAGGCAGTAGAGAATCTATTGTCCCATGACTTGACTTGGATAAAATGGAGTGTAGACTTGCAATAACTAATAAACCTGAGTCCCCACTATTTCTTTACCAGCTTTGCTTAATTACTATTCTTTATAATCATTTGTGTAGACTTGGGGGAAACATATGAAAGGGTCTGGTCCTGGAAGTGAAAGAAACACTATTTGTGAACCTTCTGCCATCATGTATTATCACTTATAATTCTACTTAAAGTGTTATAAAAAGTTACTTGTGTTTTTACATCTTATTTACTATAGCTTATTACACTTAGAAAGTCATCAGGGATTTTTCCCAGCTTCAAAGGCAAGGGCCTTAAATAATAAATTTCCCAAAGAACGCAAGCAGGGTGAGTTGGTACTATCAAAGTGGGAAGGGCTCTAGCAGGATGTGGAATTGCTATCTTGAGGAATACTGAATGCAAGCAAGGAGAATTGGTTCCTGTAATAGGAGACCCTGAAGCTGACACTGCTATTTAAACCAGGAATCATCTTACTCACCAAAATGAATCAGTACAGAAAGAGAAGGGTGTAATTTGGTCTTCCTTGTTTCTTATTTGATGAACAGGCATGGAAAGGGTTATTTATCATCTAATTGACTCAAAAATTAGAAACTAAATAGGAAAGAGGTGGGGGACAGGCCTGGTTGACCTACGAAAGACTGGCCTCTATCATGTGGGAGACAAAAGGCCAGGGACTTTTTGGCAGAGAAACAGGATTTGTGATTGGGAATATTGCTTTGCCTGTCTTCACTTGCAATAGTGCTGATGATGATGCAGGAGAAGATAGGGAGACCCCAGGTCTTGGAGCTGCCTTATTAATTTTCCCTATTAATTATCCCAACACCAGCTCCTTTTCTGCTCTCCTTAAACAGAGATTGCTTCCTTTGAAATCCTATCACCTTAGTCATTTTTATAGCCTTTTCTTGAAAGAGAGAATCACATTTCTTGGATCACTCTTAAGCATCTGTGAGTGACTGAACAGTATTCAACCCTGTTTATACCACAGAGCCTAGCATATGCATATATAATGATATTGGCTCAAATAAATATTTACCTGCTGGTCGGCTTTGGTGACACATTAGAAGCAGTCAGTCATGGTATGTTTTATGCTATTGTAAAGGAGTATAGCTACTATTTATTTCAAGGGAGGTTTGAATTTAAAGATCTTGGTAGCATAAATCCGATCTAGCAATTTGCCTCAGTTTACCTGTCCATTATCATCCAGGAGATGTCTGTATTCAGGTGGGATAGTGTCATCTCTTTCTCCAAGCTTGTCTCTATGTTCAGCTTTAGCTTTTTCCTATATCAACAGTGAGCAGATAGAATTAATATTCATTTCCTCTTTCACTTAGAAAATACATTGTCAAATGCAGCATTCAGCATTTGTTGTACTTTCATGAGGCAAAACATATGGTCTGTTTTTATTTTTTAAGCAACAGAACAACACAGATGACTTCAACAGACTGCTAGTTGGCTAAAATATAAAATCCCGTATGCTTCTGTATGTCAATTCATATCTGTGAATTTTCTAGCTATATTTTAAATGGAAATAAATGATTAAATAATTATCTTCAGAAACCATGTTAGGAGATTAGAACCCAAAGGTATAAAATATCTTTTTCTTTTTTTCTGTATGGTTTCACTTTTCTAATACAAAATCAGGCCACTGTACCTTGCCATTTAGAGAGGTCACATTTACAACTTTGCCTATTTATAAGCAACCCTGAAGGACAACAGCTAATTTGAATGGATGTGCTGACTGCTTGTGTTGCATGGGAGGAAGCCATGCATACCCCACACCTACCTCCCAGAATCCCCTCAGGGAGGCTCCGCTGGTATCTCTGTCACAGATTCAGTGACCTTCCCCTCCAGTGGAGTACAGATTGATTTTTCTATTTCATTATAATTTCATGTTTAAAATATAGGTTAAGTTCACATCAATATTCCTATGACAATGACAGAGTCAAGACCCAAGGATTAAAATTTCACTATTAGTGAATTTTTTTTTTTTTTTTTTTTTTTTTTACCTTTACTTTATCTTCCATTGGTTTGTTCTCATCTGGGTCAGGCTGCCTTTGTCCTAGACTGTCAGAGAGTTTATCCAAGGCATCATCGAGGTCTTTGTCACTCTGCTGAAAAGTAAATAATGCTGAATTAGTCACTCATTAGCCAAACTGTATCAGAGGCAAATTAGCCCAGGAACTCCATCTTTTCAGGAGGGAACCTGTTTCCTTGGAAGAAAAAGACATCTGGGGCTGGGGCAGGGGAGTAGACAGGGTCAGAGAAGAGAAGCCTAGGAATGGAAGATCAGGAGAAGAGCGGCAGTATCTGTCACCCTACTGGGGTTGGAGGGCCAGCACCTTCCACCCAACCCTGCCCATCTCTTGGTGAAGATCCCACCAGGTTAAGGAGGTCTTAGTGGCAGCCTCAGGAGCCATATCCAGTGGGTGACCTGGAGGTCACATAAAGGGTCCAAAAGCAAATGAACCAATCATGTGTGCCTTTCATTTAGAAATTAAACACCATTAGAAAACTGGATATGAAAACAAACATCTACTAATGTTGTCAGATGGTTAGGAAGCAAGATTCTGCAACTATAGAGGGTAAGTGTTTCTTTGGTTCTGTGGGTCCTCTCTAAAACTCTAAGATCTTGAGGGGTGCATTTCAGAAAGTGCAGCGTGACCCGCAGTTTGTGGGAAGCCATGGAGCTCGGCACTGCCATCCTAATACTTCCTAAAGCACAAAACCCCAGAGACAATCTGGGGTCAGGAGAGTGGAAGGGGCTTGTCTGCCACACTGGTGATGAGTGCCCTGAAAGACTTCAGAGAATTTCTGAACTGGTGGGGAAACCTCTCTTTTCATCTTCAGGAAGCTCATGGAAGTGAAATTGCAGAAATGGGAGCTGGTATTCTAGAGGAAAAAAATTATGGACAACAATATCACTGTAACTAAGATAGCTTATTTCCTCTAACATTTATTTACTGTATGATTCAGGCAGCTTATTTAACCCCTTTCAGCTTCAGTTTCCTTGGCCGTGAAATGTGAATAATAGTAGTACTTATACTCCTAAGTTGCTGAGAAAAGTAAATGATTGAAAAGGCATTTAAAACAATACTAGTTGTATGTTAGGACCCAACAAATGGTAAATTATTATTAGTATTATTATAGCAAAATCCATATTTTTCAACACATTGCATTCAAAATTCCACCTCTAAATGAATTCAATTAAAATGTGTTTAATATCTACATTGTATAAGACACCATGCTGAACTCTGTCCAATACTGTATAGAACTTTCCAGTTGATTTTCAAAATGTTTTCACATACACTATCCAGTTTTATTTGATGCCCACAATGGTTCTCAGTGAACTAAGCAGGCTTTTTTTTTTTGAGGTAAAAAGCTCAGGGAAGCTAAGTCAGTTGCTTGAAAACAAATTGTTAAAAAGTGAATAGAACCCAGGTCTTGGGACTGATAAAGCTTTTCCCTGTCATGCTTAGTCACATCCATGATCTTCTATTTTCTTTGAAGCAGTTTTCCTGTTGGAGTGATTTTATTACACAGATCTTTGAAATCATGTCTTCAAATGCTTTCAGTGTATGTAACACTGTTAGTAACAATCTAATAATCACAGCAAAGAAAGCTCCCGTGAATTATCATGGTTTATTTGACTCTTCGATTTCCTAATATTTTTATCTAAATAAAGCTTTATACCTTGTTTTAGTGACATCTTCAAATAAAATGTTAACTAAAAACAAGCCTCTCTGATGGGAAATGTGATCAGAGAAGTGTCATTGTAAAACCTACTTCTTAAAGGCAAAAAAGTTTTTGATTGCAAATGTTTACTGATAGCCTTCATCAGGGCAGAATCTCTGGCCTGAATATTAAGAACTGAAGTGTAAACGGCAGCCTAGGCTATTAATGATTCTTCCTTTCTGTTGCATGGGGACTTTCTTCATTGTGGGTGTGTTTACATACACACATGCACATGCACATGCACATACACACACACGGGGCATTTTACTGGTTTTAAGTGCTTTATTATAATCCAGGATTATAGCTGCTAATGGTAGAGCTGCCCGGGGCCAGGTCTGGGCTTTGTCATTTGTGCCTCTGGATATTTTCAG|ATGATCCTGAAGCTGACGCAACAGGATGAAAATCCATCAGAATCTCAGACTACAGCACTAAATATGCTTTGATGCTACATCAAACGGAATGGAAGCATAGCTGACTTCGCTAAAGTTACTTCATCTCCATCTAGCAAATGAGGCACTGTTCTCAACCAAAGGAGATGGGGATCTGGTTTAGGGCAATCCCTTTATAATTTGATGTGCTGTGGTCTCCTTGGTAATGTATAATTTGGTATTGCACAGGTGATTAGTCAAGGAAGTCTGGAAAAGCTTTGGTCCCACAGCCTTGCCTCACAGCATGTAAATAATTAAAACAATATTGATGCTGAGGTTCTTCTACTGCTAGTATGAAAGTGACAAATTTTTACTGGTGTGAATTGGGAAGAAAACAATGCTATTCCATGACGTTTGTAAAATGTTTGTAAAAGCTCAAACATGACGATTCCATAAAATAAACTTGAGGTTAAATAATGGGTAGTAAATTATAGAATGTAT|AAGAAAAAATATAAAGGAGAAAATCAATTATCAGGAAAGCTAAAGAACTTTTCAAATCTAGTAATTTGAATATAGACACAATGCACTTTATTGCACTTTCAATTCTTATAAAGCAACAATAATATTAAGGTCCTTGACTATGTGTACAATGTTTTCACATATATAGTTTCATTTAATCATTTCAAAGTTAATCTCTGCCATCTCGCTAAATCATCAGTCTCGGCTCTTCTGAAATAGAAGGTGCCTGATCTTCCTAATAATTCTGCCTATTTTCATTTGCTTTAAACAGGCGCCCTATTTTCTTTCTAGTTGTGGCTGCGCAAAAACATTTATCTCCCAAATAAGATGTGCTGCTTACCGAGGTATCACGGGGTGGGGCTCCAGCTTGGGTCGTTGAAGCTGGGGTTTGGGAAACCACTTCAGAGATGGCAGCAGCAAGTTTAGCATCTTCAAATTTCTTTTATTGAAAAAAATTTTATTAGTAACATGTTGTATATAAAATTATGAGCACAATGCCATCACTTAACTATAACTCTTAAAGATAGCTTAATGACTGTTTATTCTCTTGACCAAATAGACTCATAATAACATATAATTTTAAAAGAAATTTAAATTCTTTCTTCTCTATTGTATTATTTTATACAATTTGCTATTTCTATTTCCTTCTCATATTGATTATTCTAAATACTATGCAATAATATAACTTAGAGTTCCACGGTTTGTTTACACATTTCCTGTTGTACATTTAGGTTATTCAAAGTTTTCAGCTCTTTTAAAATTGCTCTGAATAAGTTCTAGTGAGTGAGTTATGGTGCTGGCTATATTTTGCTAAACTGCCCTCTCAAATGTTGCTAGGAATTCATACTGCGAAAAGCAATGAATAAGCATGCCTGTTTTCCCATGGCCTTGCTTGCCAGAATTTGACTTTTATTATGATAATCAGTGTAAAATGATATACTACTATTGCTTGTATATTGTGGTATACGGTGTCAGGTTTCAGGGTTTTTTTTCAACGTTAAATATTCTAGAAACTTTCTGAAATAATTTCTGTTTAAAAATATTGAATATTTGCTTCATTTCAAATACTCCCTTTTGACAAAAAAACTTAGGTATAACTGTTGATGAAAAACCAGAAAAAAGTCCAGAACTCTTTGGTGACTCCAACTATGGATAGCTTATTTTGAAAAAGGAGAATTGCAAATTTTACCAAAAGATGGAGAAAAGCACATTAAAAAGATACCAACATTCAGAAATTCATTTCAGCATGTTATTATTGGAAATTATTTAAACTAATTTAGATAACTATAAGATACTTATTGTCCATTTATACCCTGTAAAGCCGTTTTAGAATGTAATATTTTAGGTAATCCAAAATGTACTAAATTAAATTCATTTTTAGTTATGAGAAATCTTTGCTTATATGACAAATGAAAAGAATAACAAGTTGTCAAATGAAAAGAATGACATTGAAACATTTGTATTGTCTCTTCTTAAACTATCTTATTGACTTATTATTTAAGCCTTTTAATACTAAGTATGAAACAACCTATGGTCTGGAAATTTGTATCGCAAAGCTATATGTGCATATGTTATTTAATTCATCTAATGCTACACAAAAGCATAAAATAATGATTTTTCACTCTCTTTAAAAATACTAAATCATTTATGTCCATTTCTCAATTTTTTCATTGATCTATGCTTTGAGTTTGCTTTCTCAACATTATTGTATTTTCCACTTATTATTACTGTATAACATATGCTAGTGTTTAGTTGGATTAATCTTACCTAAAAGTACTGAAAAATGCTTTTTAGTACTTTTTCATATTTTATACATTTATTTTCCGAATGTATCATTGAATAATTTTATTGAGTTATAAAAGTATCTTATTGCTATTTAATAAAAAATTAACACATAAAATGACTTGAATTGTCATCATTCTTTTTAAGATATTTAGTTAAACTGACTTAATGTATGGCCTTCAATTTTTTTGTGTCCTTATTTTTCTGATCATTTCTCCTTTTATAGTTTACATTAAGTCTGATCTCATATTAATTACATTTTCTCATCTGTTGTTACTAATAAACATGGCATAATGTTACTTACAAATGTATTATCTACAAGTAGTGCTATCCACAAATATATTCAAATGTTCCCTTTTAATGTTTGTCATTTTTTTCATGTGTTGTTAATGATTCTTCCATGTG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5273958f-6314-4ef8-afaf-8e1d5adacc54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-691"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "offset = -len(erap_reference.split('|')[0])\n",
    "offset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "06a9a534-a67e-4bf9-b80a-b17db50e5d20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 4]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_occurrences(s, ch):\n",
    "    return [i for i, letter in enumerate(s) if letter == ch]\n",
    "\n",
    "def replace_char(s, i, c):\n",
    "    return s[:i] + c + s[i+1:]\n",
    "\n",
    "find_occurrences('AATAT', 'T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a32ab454-e8de-44c7-8b5b-f29fa8313a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_unique = {row['Sequence'].strip() for row in rows}\n",
    "sequences2i = {sequence: str(i) for i, sequence in enumerate(sequences_unique)}\n",
    "\n",
    "seqs_to_align = [\n",
    "    str_to_fasta_seqrecord(\n",
    "        str(i),\n",
    "        sequence \n",
    "    )\n",
    "    for sequence, i\n",
    "    in sequences2i.items()\n",
    "]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "42af6fa7-8945-4ae8-a7a1-a2a4d748103c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tgsts.sequtils.fastio import write_fasta\n",
    "from tgsts.sequtils import str_to_fasta_seqrecord\n",
    "from tgsts.sequtils import clean_sequence\n",
    "\n",
    "write_fasta(\n",
    "    'erap_seqs.fasta',\n",
    "    seqs_to_align,\n",
    "    force=True\n",
    ")\n",
    "\n",
    "write_fasta(\n",
    "    'erap_reference.fasta',\n",
    "    str_to_fasta_seqrecord('ref', clean_sequence(erap_reference)),\n",
    "    force=True\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9ffec571-29d3-4943-a38c-1cb36710057c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.005*2.16] collected minimizers\n",
      "[M::mm_idx_gen::0.008*2.39] sorted minimizers\n",
      "[M::main::0.008*2.38] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.009*2.26] mid_occ = 43\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.009*2.17] distinct minimizers: 8894 (98.97% are singletons); average occurrences: 1.025; average spacing: 5.278; total length: 48105\n",
      "[M::worker_pipeline::2.059*2.95] mapped 437 sequences\n",
      "[M::main] Version: 2.22-r1101\n",
      "[M::main] CMD: minimap2 -a erap_reference.fasta erap_seqs.fasta\n",
      "[M::main] Real time: 2.066 sec; CPU: 6.079 sec; Peak RSS: 0.166 GB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['minimap2', '-a', 'erap_reference.fasta', 'erap_seqs.fasta'], returncode=0)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Run minimap\n",
    "import subprocess\n",
    "\n",
    "subprocess.run(\n",
    "    [\n",
    "        \"minimap2\",\n",
    "        \"-a\",\n",
    "        \"erap_reference.fasta\",\n",
    "        \"erap_seqs.fasta\"\n",
    "    ],\n",
    "    stdout=open(\"erap_frags_aligned.sam\", \"w\"),\n",
    "    check=True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d631bf4e-2be7-40a1-9120-ee5f5e2c8676",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load aligned\n",
    "\n",
    "from utils import build_full_length_alignments\n",
    "\n",
    "ref_aln, subread_strings = build_full_length_alignments('erap_frags_aligned.sam', 'erap_reference.fasta')\n",
    "\n",
    "piped_ref_aln, piped_read_alns_strs = add_pipes(erap_reference, ref_aln, subread_strings)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "27129db0-ece9-4ebc-9b28-3abf90d6afa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Alignment:\n",
    "    def __init__(\n",
    "        self,\n",
    "        ref: str,\n",
    "        ali: str\n",
    "    ):\n",
    "        self.ref = ref\n",
    "        self.ali = ali\n",
    "\n",
    "        self.rrs = identify_rrs_ali(ali)\n",
    "        self.seq_rr_masked = mask_rrs(clean_sequence(ali))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b0389752-e44d-4291-9971-6ed243ee8b76",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing: 100%|| 437/437 [01:38<00:00,  4.45it/s]\n"
     ]
    }
   ],
   "source": [
    "#Generate alignments\n",
    "def generate_alignment(key, ali):\n",
    "    return Alignment(piped_ref_aln, ali)\n",
    "\n",
    "args = [\n",
    "    (k, ali)\n",
    "    for k, ali\n",
    "    in piped_read_alns_strs.items()\n",
    "]\n",
    "\n",
    "res = run_concurrently(\n",
    "    generate_alignment,\n",
    "    args,\n",
    "    mode='process',\n",
    "    batch_size=10\n",
    ")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ae732c0f-00ff-4d75-bb7c-33dad9d0165b",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert not res.failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ce998873-a700-4550-96b2-6017ab88301b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "piped_read_alns = {\n",
    "    r.args[0]: r.result\n",
    "    for r\n",
    "    in res.successful\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "71d5baba-5fb1-488e-bdcf-6adfca18f3cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(iter(piped_read_alns.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c8773b2f-7646-4d00-b757-ff4a8e4bc3a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_erap_exonic(gen_seq: str) -> str:\n",
    "    ea = exonic_alignment(\n",
    "        str_to_fasta_seqrecord('cons', gen_seq),\n",
    "        'ERAP1',\n",
    "        'erap_full'\n",
    "    )\n",
    "    return ExonDict(*next(iter(ea.items()))).seq\n",
    "\n",
    "def calculate_mms(seq: str, mode: str = 'gen') -> str:\n",
    "    seq = str_to_fasta_seqrecord('cons', clean_sequence(seq))\n",
    "    locus_lib = ANTypingLibs().get_locus_lib('ERAP1')\n",
    "    locus_lib.load_db()\n",
    "\n",
    "    if mode == 'gen':\n",
    "        mms, _ = mm_profile_from_seqs(\n",
    "            seq,\n",
    "            locus_lib.get_gen_ref_seq(), \n",
    "            locus_lib.get_gen_ref_seq(), \n",
    "            locus_lib,\n",
    "            'gen'\n",
    "        )\n",
    "    elif mode == 'cds':\n",
    "        mms, _ = mm_profile_from_seqs(\n",
    "            seq,\n",
    "            locus_lib.get_cds_ref_seq(),\n",
    "            locus_lib.get_cds_ref_seq(),\n",
    "            locus_lib,\n",
    "            'cds'\n",
    "        )\n",
    "        \n",
    "    else:\n",
    "        raise ValueError('mode must be gen or cds')\n",
    "    mms = ', '.join([str(mm) for mm in mms])\n",
    "    return mms\n",
    "        \n",
    "    \n",
    "def get_erap_cds_prot(gen_str: str) -> str:\n",
    "    annotator = Annotator('gen')\n",
    "    res = annotator('ace78ee52a044de361c280b47d9e8', gen_str, fast_alignment=True)\n",
    "    return res['sequence']['coding'], res['sequence']['protein']\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "20053902-f587-4465-95dc-324d693e5a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "class OverlapError(Exception):\n",
    "    def __init__(\n",
    "        self,\n",
    "        message,\n",
    "        sample,\n",
    "        frags_left,\n",
    "        frags_right,\n",
    "        overlaps\n",
    "    ):\n",
    "        super().__init__(message)\n",
    "        self.message = message\n",
    "        self.sample = sample\n",
    "        self.frags_left = frags_left\n",
    "        self.frags_right = frags_right\n",
    "        self.overlaps = overlaps\n",
    "\n",
    "        self.write_output()\n",
    "\n",
    "    def write_output(self):\n",
    "        out_path = f'./out/fails/{self.sample}.txt'\n",
    "\n",
    "        if os.path.exists(out_path):\n",
    "            os.remove(out_path)\n",
    "\n",
    "        with open(out_path, 'w') as f:\n",
    "            f.write(self.sample+ '\\n')\n",
    "            f.write(self.message + '\\n\\n')\n",
    "\n",
    "        if_left = self.frags_left[0].fragment\n",
    "        if_right = self.frags_right[0].fragment\n",
    "\n",
    "        overlap_lines = []\n",
    "        overlap_lines.append(f'         \\tCoding\\tNonCod\\tRR    \\tKmer')\n",
    "        for (i, j), overlap in self.overlaps.items():\n",
    "            overlap_lines.append('\\t'.join([\n",
    "                f'F{if_left}:{i}/F{if_right}:{j}',\n",
    "                *[\n",
    "                    str(int(_)).ljust(6)\n",
    "                    for _\n",
    "                    in overlap.match_tuple\n",
    "                ]\n",
    "            ]))\n",
    "\n",
    "        with open(out_path, 'a') as f:\n",
    "            f.write('\\n')\n",
    "            f.write('\\n'.join(overlap_lines) + '\\n\\n')\n",
    "                    \n",
    "\n",
    "        ali_lines = []\n",
    "        msa = MSA(*self.frags_left, *self.frags_right)\n",
    "\n",
    "        #Get overlapping boundary idx\n",
    "        i_start = max([\n",
    "            len(re.match(r'[-|]*', ali).group())\n",
    "            for ali\n",
    "            in msa.alis \n",
    "        ])\n",
    "        i_end = len(msa.ref) - max([\n",
    "            len(re.match(r'[-|]*', ali[::-1]).group())\n",
    "            for ali\n",
    "            in msa.alis \n",
    "        ])\n",
    "\n",
    "        ref = msa.ref[i_start:i_end]\n",
    "        alis = [ref] + [ali[i_start:i_end] for ali in msa.alis]\n",
    "\n",
    "        #Lose all columns which are just -s\n",
    "        ref, *(alis) = [\n",
    "            ''.join(ali)\n",
    "            for ali\n",
    "            in zip(*[\n",
    "                bases\n",
    "                for bases\n",
    "                in zip(*alis)\n",
    "                if set(bases) != {'-'}\n",
    "            ])\n",
    "        ]\n",
    "        \n",
    "        labels = [\n",
    "            f'F{if_left}:1', \n",
    "            f'F{if_left}:2', \n",
    "            f'F{if_right}:1', \n",
    "            f'F{if_right}:2', \n",
    "        ]\n",
    "        batch_width = 80\n",
    "\n",
    "        for i in range(0, len(ref), batch_width):\n",
    "            s = slice(i,i+batch_width)\n",
    "\n",
    "            ref_snippet = ref[s]\n",
    "            alis_snippets = [ali[s] for ali in alis]\n",
    "\n",
    "            label_ljust = 20\n",
    "            \n",
    "            mm_line = ''.join([' '] * label_ljust) + ''.join([\n",
    "                'v' if len(set(bases)) != 1 else ' '\n",
    "                for bases\n",
    "                in zip(*alis_snippets)\n",
    "            ])\n",
    "\n",
    "            ali_lines.append(mm_line)\n",
    "            ali_lines.append('ref'.ljust(label_ljust) + ref_snippet)\n",
    "            ali_lines.extend([\n",
    "                label.ljust(label_ljust) + ali\n",
    "                for label, ali\n",
    "                in zip(labels, alis_snippets)\n",
    "            ])\n",
    "\n",
    "            \n",
    "\n",
    "        with open(out_path, 'a') as f:\n",
    "            f.write('\\n'.join(ali_lines))\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "767f744b-1cfb-4eb8-9083-684f5611e16b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class AssemblyError(Exception):\n",
    "    def __init__(self, msg: str, sample: str, overlap_lls: list['OverlapLinkedList']):\n",
    "        super().__init__(msg)\n",
    "        self.sample = sample\n",
    "\n",
    "        self.write_output(msg, overlap_lls)\n",
    "\n",
    "    def write_output(self, msg: str, overlap_lls: list['OverlapLinkedList']):\n",
    "        \n",
    "        out_path = f'./out/fails/{self.sample}.txt'\n",
    "\n",
    "        lines = []\n",
    "\n",
    "        lines.append('SAMPLE:')\n",
    "        lines.append(self.sample)\n",
    "        lines.append('REASON:')\n",
    "        lines.append(msg)\n",
    "\n",
    "        #Left to right overlaps, i.e 4 frag_1/2\n",
    "        l2r_overlaps = {}\n",
    "        \n",
    "        for overlap_ll in overlap_lls:\n",
    "            for overlap in overlap_ll:\n",
    "                l2r_overlaps.setdefault(\n",
    "                    (overlap.frag_left.fragment, overlap.frag_right.fragment),\n",
    "                    set()\n",
    "                ).add(overlap)\n",
    "\n",
    "        lines.append('\\n')\n",
    "\n",
    "        for (i_frag_left, i_frag_right), overlaps in l2r_overlaps.items():\n",
    "\n",
    "            \n",
    "            frags_left = sorted({\n",
    "                overlap.frag_left\n",
    "                for overlap\n",
    "                in overlaps\n",
    "            }, key = lambda x: id(x))\n",
    "            frags_right = sorted({\n",
    "                overlap.frag_right\n",
    "                for overlap\n",
    "                in overlaps\n",
    "            }, key = lambda x: id(x))\n",
    "\n",
    "            #Get common mismatches across all frags\n",
    "            assert len(frags_left) == 2\n",
    "\n",
    "            left_mms_common = set(frags_left[0].gen_mms) & set(frags_left[1].gen_mms)\n",
    "\n",
    "            lines.append(''*79)\n",
    "            lines.append(f' FRAG{i_frag_left} '.ljust(48, '').rjust(79, ''))\n",
    "            lines.append(''*79)\n",
    "            lines.append('')\n",
    "            \n",
    "            for i, frag in enumerate(frags_left):\n",
    "                lines.append(f'ALLELE {i+1} UNIQUE MISMATCHES:')\n",
    "                lines.append('\\n'.join(sorted(set(frag.gen_mms) - left_mms_common, key=mm2pos)) or 'NONE')\n",
    "                lines.append('')\n",
    "            lines.append('')\n",
    "\n",
    "            \n",
    "            lines.append(''*79)\n",
    "            lines.append(f' OVLP{i_frag_left}/{i_frag_right} '.ljust(50, '').rjust(79, ''))\n",
    "            lines.append(''*79)\n",
    "            lines.append('')\n",
    "\n",
    "            lines.append('OVERLAP STATUS:')\n",
    "            if len(overlaps) == 2:\n",
    "                lines.append('UNAMBIGUOUS OVERLAP')\n",
    "                lines.append('')\n",
    "            else:\n",
    "                overlap_alignments = [\n",
    "                    overlap.get_alignment()\n",
    "                    for overlap\n",
    "                    in overlaps\n",
    "                ]\n",
    "                unique_overlap_count = len(set(overlap_alignments))\n",
    "                if unique_overlap_count == 1:\n",
    "                    lines.append('AMBIGUOUS HOMOZYGOUS OVERLAP')\n",
    "                    lines.append('')\n",
    "                else:\n",
    "                    lines.append('AMBIGUOUS HETEROZYGOUS OVERLAP')\n",
    "                    lines.append('')\n",
    "                    \n",
    "            lines.append(' '.join(['Overlap'.ljust(20), 'C', 'G', 'R', 'D']))\n",
    "            for i_allele_left, frag_left in enumerate(frags_left):\n",
    "                for i_allele_right, frag_right in enumerate(frags_right):\n",
    "\n",
    "                    \n",
    "                    for overlap in overlaps:\n",
    "                        if (\n",
    "                            overlap.frag_left is frag_left and\n",
    "                            overlap.frag_right is frag_right \n",
    "                        ):\n",
    "                            label_left = f'F{i_frag_left}A{i_allele_left}'\n",
    "                            label_right = f'F{i_frag_right}A{i_allele_right}'\n",
    "                            \n",
    "                            #print(f'{label_left} / {label_right} Overlap Alignment')\n",
    "                            #alignment = overlap.get_alignment(padding=20)\n",
    "                            #ref, alis = alignment[0], alignment[1:]\n",
    "                            #for i_ali_start in range(0, len(ref), 69):\n",
    "                            #    i_ali_end  = i_ali_start + 69\n",
    "                            #    print('ref'.ljust(10), ref[i_ali_start:i_ali_end], sep='')\n",
    "                            #    for label, ali in zip((label_left, label_right), alis):\n",
    "                            #        print(label.ljust(10), ali[i_ali_start:i_ali_end], sep='')\n",
    "                            #    print()\n",
    "\n",
    "                            lines.append(' '.join(\n",
    "                                map(\n",
    "                                    str,\n",
    "                                    [\n",
    "                                        f'{label_left}/{label_right}'.ljust(20), \n",
    "                                        *map(int, overlap.match_tuple)\n",
    "                                    ]    \n",
    "                                )\n",
    "                            ))\n",
    "            lines.append('')\n",
    "            lines.append('')\n",
    "\n",
    "\n",
    "        #Now do the final right frag\n",
    "        right_mms_common = set(frags_right[0].gen_mms) & set(frags_right[1].gen_mms)\n",
    "\n",
    "        lines.append(''*79)\n",
    "        lines.append(f' FRAG{i_frag_right} '.ljust(48, '').rjust(79, ''))\n",
    "        lines.append(''*79)\n",
    "        lines.append('')\n",
    "        \n",
    "        for i, frag in enumerate(frags_right):\n",
    "            lines.append(f'ALLELE {i+1} UNIQUE MISMATCHES:')\n",
    "            lines.append('\\n'.join(sorted(set(frag.gen_mms) - right_mms_common, key=mm2pos)) or 'NONE')\n",
    "            lines.append('')\n",
    "        lines.append('')\n",
    "\n",
    "        with open(out_path, 'w') as f:\n",
    "            for line in lines:\n",
    "                f.write(line + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cd75174-60c5-4b0d-a780-c565c3937cb7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b131471-bec3-4be4-a2c8-8d152be75f43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a68946a3-08c8-40c4-a74e-fb66e2dd2a68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-296"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class Fragment:\n",
    "    def __init__(\n",
    "        self, \n",
    "        seq: str,\n",
    "        ali: Alignment,\n",
    "        fragment: int,\n",
    "        read_count: int,\n",
    "        fragment_name: str,\n",
    "        gen_mms: list[str]\n",
    "    ):\n",
    "        self.seq = seq\n",
    "        self.ali = ali\n",
    "        self.fragment = fragment\n",
    "        self.read_count = read_count\n",
    "        self.fragment_name = fragment_name\n",
    "        self.gen_mms = gen_mms\n",
    "\n",
    "        try:\n",
    "            self.mm_pos_dict = {\n",
    "                mm2pos(mm): mm\n",
    "                for mm\n",
    "                in gen_mms\n",
    "            }\n",
    "        except:\n",
    "            print(gen_mms)\n",
    "            raise\n",
    "    \n",
    "    def merge_aln(self, other: 'Fragment'):\n",
    "        return MSA(self, other)\n",
    "\n",
    "def mm2pos(mm: str):\n",
    "    return int(re.search(r':(-?\\d+)', mm).group(1))\n",
    "\n",
    "mm2pos('5utr:-296delACACACACACACAC>')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "23bc5eb4-f896-432c-bb8a-1b9072405bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Overlap:\n",
    "    def __init__(\n",
    "        self,\n",
    "        frag_left: Fragment,\n",
    "        frag_right: Fragment,\n",
    "    ):\n",
    "        self.frag_left = frag_left\n",
    "        self.frag_right = frag_right\n",
    "\n",
    "        self.match_tuple = self.assess_overlap()\n",
    "\n",
    "    def __bool__(self):\n",
    "        return self.match_tuple[0] and self.match_tuple[2]\n",
    "\n",
    "    def __str__(self):\n",
    "        return str(self.match_tuple)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'Overlap{self.match_tuple}'\n",
    "\n",
    "    def get_alignment(self, padding: int = 0):\n",
    "        \"\"\"\n",
    "        Return alignment tuple of the overlap between frag left and right\n",
    "        \"\"\"\n",
    "        overlap_aln = self.frag_left.merge_aln(self.frag_right)\n",
    "        return overlap_aln.get_trimmed_alignment(padding=padding)\n",
    "\n",
    "\n",
    "    def assess_overlap(self):\n",
    "        overlap_aln = self.frag_left.merge_aln(self.frag_right)\n",
    "\n",
    "        #overlap_aln.print()\n",
    "\n",
    "        ref_ali = overlap_aln.ref\n",
    "        alis = overlap_aln.alis\n",
    "\n",
    "        i_start, i_end = overlap_aln.get_alignment_overlap_idx()\n",
    "\n",
    "        #Get union of rr bases\n",
    "        rrs1, rrs2 = overlap_aln.rrs\n",
    "        rr_bases_1 = {\n",
    "            i\n",
    "            for rr\n",
    "            in rrs1\n",
    "            for i\n",
    "            in range(rr['ali_start'], rr['ali_end'])\n",
    "        }\n",
    "        rr_bases_2 = {\n",
    "            i\n",
    "            for rr\n",
    "            in rrs2\n",
    "            for i\n",
    "            in range(rr['ali_start'], rr['ali_end'])\n",
    "        }\n",
    "        rr_bases_both = rr_bases_1 | rr_bases_2\n",
    "\n",
    "        #Set to false when mismatch found\n",
    "        coding_match, noncoding_match, rr_match = True, True, True\n",
    "        \n",
    "        for i, (ref_base, *(ali_bases)) in enumerate(zip(ref_ali, zip(*alis))):\n",
    "            if not i_start <= i < i_end:\n",
    "                continue\n",
    "            if len(set(*ali_bases)) == 1:\n",
    "                continue\n",
    "            #Mismtch found\n",
    "            #print(i, ref_base, ali_bases, overlap_aln.get_feature(i), end='\\t')\n",
    "            feature = overlap_aln.get_feature(i)\n",
    "            if feature == 'coding':\n",
    "                coding_match = False\n",
    "            elif feature == 'noncoding':\n",
    "                noncoding_match = False\n",
    "            else:\n",
    "                raise ValueError('unexpected feature found?')\n",
    "\n",
    "        left_tail_masked, right_head_masked = trim_intersection(\n",
    "            self.frag_left.ali.seq_rr_masked,\n",
    "            self.frag_right.ali.seq_rr_masked,\n",
    "        )\n",
    "        if len({\n",
    "            str(left_tail_masked.seq),\n",
    "            str(right_head_masked.seq)\n",
    "        }) != 1:\n",
    "            rr_match = False\n",
    "\n",
    "        kmer_similarity = - calculate_kmer_distance(\n",
    "            clean_sequence(alis[0][i_start:i_end]),\n",
    "            clean_sequence(alis[1][i_start:i_end]),\n",
    "            7\n",
    "        )\n",
    "\n",
    "        return coding_match, noncoding_match, rr_match, kmer_similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3879c220-94a9-4423-a11d-3991f2245b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MSA:\n",
    "    def __init__(\n",
    "        self,\n",
    "        *fragments: Fragment\n",
    "    ):\n",
    "\n",
    "        if len({\n",
    "            f.ali.ref\n",
    "            for f\n",
    "            in fragments\n",
    "        }) != 1:\n",
    "            raise ValueError('ref must be identical')\n",
    "\n",
    "        self.fragments = fragments\n",
    "    \n",
    "        self.ref = fragments[0].ali.ref\n",
    "        self.alis = [f.ali.ali for f in fragments]\n",
    "        self.rrs = [f.ali.rrs for f in fragments]\n",
    "\n",
    "    @classmethod\n",
    "    def from_fragments(cls, *fragments: Fragment):\n",
    "        return cls.from_alis(*[\n",
    "            frag.ali\n",
    "            for frag\n",
    "            in fragments\n",
    "        ])\n",
    "            \n",
    "\n",
    "    def get_feature(self, i):\n",
    "        upstream = self.ref[:i]\n",
    "        pipe_count = upstream.count('|')\n",
    "        if pipe_count % 2 == 0:\n",
    "            return 'noncoding'\n",
    "        else:\n",
    "            return 'coding'\n",
    "            \n",
    "    def get_alignment_overlap_idx(self):\n",
    "        #Get overlapping boundary idx\n",
    "        i_start = max([\n",
    "            len(re.match(r'[-|]*', ali).group())\n",
    "            for ali\n",
    "            in self.alis\n",
    "        ])\n",
    "        i_end = len(self.ref) - max([\n",
    "            len(re.match(r'[-|]*', ali[::-1]).group())\n",
    "            for ali\n",
    "            in self.alis\n",
    "        ])\n",
    "        return i_start, i_end\n",
    "\n",
    "    def get_trimmed_alignment(self, padding: int = 0):\n",
    "        i_start, i_end = self.get_alignment_overlap_idx()\n",
    "\n",
    "        i_start -= padding\n",
    "        i_end += padding\n",
    "\n",
    "        return (\n",
    "            self.ref[i_start:i_end],\n",
    "            *[\n",
    "                ali[i_start:i_end]\n",
    "                for ali\n",
    "                in self.alis\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    def consolidate(self) -> str:\n",
    "\n",
    "        #Sort by i_fragment\n",
    "        msa_sorted = MSA(*sorted(\n",
    "            self.fragments,\n",
    "            key = lambda x: x.fragment\n",
    "        ))\n",
    "\n",
    "        #Counts of each fragments\n",
    "        fragment_counts = Counter([\n",
    "            f.fragment\n",
    "            for f\n",
    "            in msa_sorted.fragments\n",
    "        ])\n",
    "\n",
    "        #Can only align if there is max 1 seq per fragment\n",
    "        #Frags must also be contiguous\n",
    "        i_fragment_prev = None\n",
    "        for i_fragment, count in fragment_counts.items():\n",
    "            if count > 1:\n",
    "                raise ValueError(\n",
    "                    'Cannot consolidate unless only 1 poss seq per fragment'\n",
    "                )\n",
    "            if (\n",
    "                i_fragment_prev \n",
    "                and (i_fragment - i_fragment_prev) > 1\n",
    "            ):\n",
    "                raise ValueError(\n",
    "                    'Fragments must be contiguous and overlapping'\n",
    "                )\n",
    "            i_fragment_prev = i_fragment\n",
    "                    \n",
    "        #Always choose fragment with higher read count\n",
    "\n",
    "        #Sort fragment indexes, higher read count first\n",
    "        prioritised_fragments: list[Fragment] = sorted(\n",
    "            msa_sorted.fragments,\n",
    "            key = lambda f: f.read_count,\n",
    "            reverse=True\n",
    "        )\n",
    "        prioritised_alignments = [\n",
    "            fragment.ali.ali\n",
    "            for fragment\n",
    "            in prioritised_fragments\n",
    "        ]\n",
    "        \n",
    "        consolidated = []\n",
    "\n",
    "        for col in range(len(prioritised_alignments[0])):\n",
    "            base = '-'\n",
    "            for i, aln in enumerate(prioritised_alignments):\n",
    "                if aln[col] != '-':\n",
    "                    base = aln[col]\n",
    "                    prev_base_frag = i\n",
    "                    break\n",
    "            consolidated.append(base)\n",
    "\n",
    "        return ''.join(consolidated)\n",
    "\n",
    "    def print(self):\n",
    "        ali_len = len(self.ref)\n",
    "        for i in range(0, ali_len, 80):\n",
    "            print(\n",
    "                self.ref[i:i+80],\n",
    "                *[\n",
    "                    ali[i:i+80]\n",
    "                    for ali\n",
    "                    in self.alis\n",
    "                ],\n",
    "                sep='\\n',\n",
    "                end='\\n\\n'\n",
    "            )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class AssembledSequence:\n",
    "    def __init__(self, *overlaps: Overlap):\n",
    "        \"\"\"\n",
    "        Assemble all relevant information from list of frags\n",
    "        \"\"\"\n",
    "        frags = [\n",
    "            overlaps[0].frag_left, \n",
    "            *[\n",
    "                overlap.frag_right\n",
    "                for overlap\n",
    "                in overlaps\n",
    "            ]\n",
    "        ]\n",
    "    \n",
    "        ali = MSA(*frags).consolidate()\n",
    "        self.seq = clean_sequence(ali)\n",
    "    \n",
    "\n",
    "        self.overlaps = overlaps\n",
    "        self.frags = frags\n",
    "        self.mms = calculate_mms(self.seq, mode='gen')\n",
    "        self.exonic = get_erap_exonic(self.seq)\n",
    "        self.cds, self.prot = get_erap_cds_prot(self.seq)\n",
    "        self.cds_mms = calculate_mms(self.exonic, mode='cds')\n",
    "        self.name = ''.join(\n",
    "            frag.fragment_name\n",
    "            for frag\n",
    "            in sorted(\n",
    "                frags,\n",
    "                key = lambda x: x.fragment\n",
    "            )\n",
    "        )\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.seq\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "75d460fb-af6d-428a-b281-35e3bbf2ef13",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FragmentPair(list):\n",
    "    def __init__(\n",
    "        self,\n",
    "        frag1: Fragment,\n",
    "        frag2: Fragment,\n",
    "        fragment: int\n",
    "    ):\n",
    "        self.frag1 = frag1\n",
    "        self.frag2 = frag2\n",
    "        self.fragment = fragment\n",
    "\n",
    "        super().__init__([frag1, frag2])\n",
    "\n",
    "    def product(self, other: 'FragmentPair', enum=False):\n",
    "        if not enumerate:\n",
    "            for a, b in product(self, other):\n",
    "                yield a, b \n",
    "        else:\n",
    "            for (i, a), (j, b) in product(\n",
    "                enumerate(self), enumerate(other)\n",
    "            ):\n",
    "                yield (i, a), (j, b)\n",
    "\n",
    "class TilingRow:\n",
    "\n",
    "    header = (\n",
    "        'sample',\n",
    "        'desc',\n",
    "        'allele',\n",
    "        'allele_name',\n",
    "        'exon_count',\n",
    "        'cds_mms',\n",
    "        'cds_mms_trunc',\n",
    "        'gen_mms',\n",
    "        'gen_mms_trunc',\n",
    "        'cds_homoz',\n",
    "        'gen_homoz',\n",
    "        '12CDS',\n",
    "        '12RRA',\n",
    "        '12GEN',\n",
    "        '23CDS',\n",
    "        '23RRA',\n",
    "        '23GEN',\n",
    "        '34CDS',\n",
    "        '34RRA',\n",
    "        '34GEN',\n",
    "        '45CDS',\n",
    "        '45RRA',\n",
    "        '45GEN',\n",
    "        'prot_seq',\n",
    "        'cds_seq',\n",
    "        'gen_seq',\n",
    "    )\n",
    "    \n",
    "    def __init__(\n",
    "        self,\n",
    "        sample: str,\n",
    "        desc: str,\n",
    "        i_allele: int,\n",
    "        allele_name: str = '',\n",
    "        cds_mms: str = '',\n",
    "        gen_mms: str = '',\n",
    "        homozygous_cds: str = '',\n",
    "        homozygous_gen: str = '',\n",
    "        prot_seq: str = '',\n",
    "        cds_seq: str = '',\n",
    "        seq: str = '',\n",
    "        overlaps: list[Overlap] = None\n",
    "    ):\n",
    "        overlap_dict = {}\n",
    "        if overlaps:\n",
    "            #Sorted overlaps for increasing frags\n",
    "            overlaps = sorted(\n",
    "                overlaps,\n",
    "                key = lambda x: x.frag_left.fragment\n",
    "            )\n",
    "            for overlap in overlaps:\n",
    "                label = f'{overlap.frag_left.fragment}{overlap.frag_right.fragment}'\n",
    "                overlap_dict[f'{label}CDS'] = 'Y' if overlap.match_tuple[0] else 'N'\n",
    "                overlap_dict[f'{label}RRA'] = 'Y' if overlap.match_tuple[2] else 'N'\n",
    "                overlap_dict[f'{label}GEN'] = 'Y' if overlap.match_tuple[1] else 'N'\n",
    "\n",
    "        exon_count = (\n",
    "            ''\n",
    "            if not seq\n",
    "            else '20'\n",
    "            if 'intron19:33433T>A'\n",
    "            in gen_mms\n",
    "            else '19'\n",
    "        )\n",
    "\n",
    "        cds_mms_trunc = ', '.join([\n",
    "            mm\n",
    "            for mm\n",
    "            in cds_mms.split(', ')\n",
    "            if mm and 'exon1:' not in mm\n",
    "        ])\n",
    "        try:\n",
    "            gen_mms_trunc = ', '.join([\n",
    "                mm\n",
    "                for mm\n",
    "                in gen_mms.split(', ')\n",
    "                if mm and mm2pos(mm) >= 3768\n",
    "            ])\n",
    "        except:\n",
    "            print(type(gen_mms))\n",
    "            print(gen_mms)\n",
    "            raise\n",
    "    \n",
    "        \n",
    "        self.tuple = tuple([\n",
    "            sample,\n",
    "            desc,\n",
    "            str(i_allele),\n",
    "            allele_name,\n",
    "            exon_count,\n",
    "            cds_mms,\n",
    "            cds_mms_trunc,\n",
    "            gen_mms,\n",
    "            gen_mms_trunc,\n",
    "            homozygous_cds,\n",
    "            homozygous_gen,\n",
    "            overlap_dict.get('12CDS', ''),\n",
    "            overlap_dict.get('12RRA', ''),\n",
    "            overlap_dict.get('12GEN', ''),\n",
    "            overlap_dict.get('23CDS', ''),\n",
    "            overlap_dict.get('23RRA', ''),\n",
    "            overlap_dict.get('23GEN', ''),\n",
    "            overlap_dict.get('34CDS', ''),\n",
    "            overlap_dict.get('34CDS', ''),\n",
    "            overlap_dict.get('34RRA', ''),\n",
    "            overlap_dict.get('45CDS', ''),\n",
    "            overlap_dict.get('45RRA', ''),\n",
    "            overlap_dict.get('45GEN', ''),\n",
    "            prot_seq,\n",
    "            cds_seq,\n",
    "            seq\n",
    "        ])\n",
    "            \n",
    "            \n",
    "\n",
    "class FailedTiling:\n",
    "    def __init__(\n",
    "        self,\n",
    "        sample: str,\n",
    "        reason: str\n",
    "    ):\n",
    "        self.sample = sample\n",
    "        self.reason = reason\n",
    "\n",
    "    @property\n",
    "    def rows(self):\n",
    "        return [\n",
    "            TilingRow(\n",
    "                self.sample,\n",
    "                self.reason,\n",
    "                i+1,\n",
    "            )\n",
    "            for i \n",
    "            in range(2)\n",
    "        ]\n",
    "            \n",
    "\n",
    "class Tiling:\n",
    "    def __init__(\n",
    "        self,\n",
    "        sample: str,\n",
    "        *fragment_pairs: FragmentPair\n",
    "    ):\n",
    "        self.sample = sample\n",
    "        self.fragment_pairs = fragment_pairs\n",
    "        self.assemblies = self.assemble()\n",
    "\n",
    "    @classmethod\n",
    "    def try_init(\n",
    "        cls,\n",
    "        sample: str,\n",
    "        *args,\n",
    "        **kwargs\n",
    "    ):\n",
    "        try:\n",
    "            return cls(sample, *args, **kwargs)\n",
    "        except (OverlapError, AssemblyError) as e:\n",
    "            print(f'{type(e)}: {str(e)}', end='\\n\\n')\n",
    "            return FailedTiling(sample, str(e)) \n",
    "\n",
    "    @property\n",
    "    def rows(self):\n",
    "\n",
    "        homozygous_cds = len({\n",
    "            assembly.cds\n",
    "            for assembly\n",
    "            in self.assemblies\n",
    "        }) == 1\n",
    "        homozygous_gen = len({\n",
    "            str(assembly)\n",
    "            for assembly\n",
    "            in self.assemblies\n",
    "        }) == 1\n",
    "        \n",
    "        rows = [\n",
    "            TilingRow(\n",
    "                self.sample,\n",
    "                '',\n",
    "                i+1,\n",
    "                assembly.name,\n",
    "                assembly.cds_mms,\n",
    "                assembly.mms,\n",
    "                'Y' if homozygous_cds else 'N',\n",
    "                'Y' if homozygous_gen else 'N',\n",
    "                assembly.prot,\n",
    "                assembly.cds,\n",
    "                str(assembly),\n",
    "                assembly.overlaps\n",
    "            )\n",
    "            for i, assembly\n",
    "            in enumerate(self.assemblies)\n",
    "        ]\n",
    "        return rows\n",
    "\n",
    "    def assemble(self):\n",
    "\n",
    "        possible_assemblies = PossibleAssemblies(self.sample)\n",
    "        \n",
    "        for frags_left, frags_right in zip(\n",
    "            self.fragment_pairs[:-1],\n",
    "            self.fragment_pairs[1:],\n",
    "        ):\n",
    "            i_f_left: int = frags_left.fragment \n",
    "            i_f_right: int = frags_right.fragment \n",
    "            overlaps = {}\n",
    "            for (i, frag_left), (j, frag_right) in frags_left.product(\n",
    "                frags_right, \n",
    "                enum=True\n",
    "            ):\n",
    "                overlap = Overlap(frag_left, frag_right)\n",
    "                #print(f'- A{i}/A{j}', overlap.match_tuple, bool(overlap))\n",
    "                overlaps[(i,j)] = overlap\n",
    "\n",
    "            acceptable_overlaps = {\n",
    "                k:v\n",
    "                for k, v\n",
    "                in overlaps.items()\n",
    "                if v\n",
    "            }\n",
    "\n",
    "            if len(acceptable_overlaps) == 0:\n",
    "                raise OverlapError(\n",
    "                    f'No overlaps found between F{i_f_left} and F{i_f_right}',\n",
    "                    self.sample,\n",
    "                    frags_left,\n",
    "                    frags_right,\n",
    "                    overlaps\n",
    "                )\n",
    "            if {_[0] for _ in acceptable_overlaps.keys()} != {0, 1}:\n",
    "                raise OverlapError(\n",
    "                    f'F{i_f_left} could not overlap both frags to F{i_f_right}',\n",
    "                    self.sample,\n",
    "                    frags_left,\n",
    "                    frags_right,\n",
    "                    overlaps\n",
    "                )\n",
    "            if {_[1] for _ in acceptable_overlaps.keys()} != {0, 1}:\n",
    "                raise OverlapError(\n",
    "                    f'F{i_f_left} could not overlap to both F{i_f_right} frags',\n",
    "                    self.sample,\n",
    "                    frags_left,\n",
    "                    frags_right,\n",
    "                    overlaps\n",
    "                )\n",
    "\n",
    "            #If there are three possible overlaps, one is always\n",
    "            #removeable as one fragment has an unambiguous choicfe\n",
    "            if len(acceptable_overlaps) == 3:\n",
    "                f_left_counts = {0: [], 1:[]}\n",
    "                for _i, _j in acceptable_overlaps.keys():\n",
    "                    f_left_counts[_i].append(_j)\n",
    "                #Find the fragment which has two options\n",
    "                greedy_frag = 0 if len(f_left_counts[0]) == 2 else 1\n",
    "                modest_frag = int(not greedy_frag)\n",
    "\n",
    "                del acceptable_overlaps[(\n",
    "                    greedy_frag, \n",
    "                    f_left_counts[modest_frag][0]\n",
    "                )]\n",
    "            #If there are 2 possible overlaps, perfect and unambig\n",
    "            #If there are 4 ambigs, this is am ambiguous overlap.\n",
    "            #This may or may not be a problem, depending on whether there\n",
    "            #are polymorphisms both upstream and downstream that become\n",
    "            #ambiguous due to the overlap ambiguity in this frag\n",
    "            #Either way, here we add all possibilities\n",
    "            \n",
    "\n",
    "            possible_assemblies.update(list(acceptable_overlaps.values()))\n",
    "            \n",
    "            #print()\n",
    "        #print(len(possible_assemblies), 'poss assemblies')\n",
    "\n",
    "        overlaps1, overlaps2 = possible_assemblies.get_final_assemblies()\n",
    "        \n",
    "        #Consolidate final sequence from the frags \n",
    "\n",
    "        assembled1 = AssembledSequence(*overlaps1)\n",
    "        assembled2 = AssembledSequence(*overlaps2)\n",
    "\n",
    "        return assembled1, assembled2\n",
    "\n",
    "            \n",
    "class OverlapLinkedList:\n",
    "    def __init__(\n",
    "        self,\n",
    "        head: Overlap,\n",
    "        tail: 'OverlapLinkedList' = None\n",
    "    ):\n",
    "        self.head = head\n",
    "        self.tail = tail\n",
    "\n",
    "    def update(\n",
    "        self,\n",
    "        new_head: Overlap\n",
    "    ) -> 'OverlapLinkedList':\n",
    "        if self.head.frag_right is not new_head.frag_left:\n",
    "            raise ValueError('New head doesnt match list')\n",
    "        return OverlapLinkedList(\n",
    "            new_head,\n",
    "            self\n",
    "        )\n",
    "\n",
    "    def __iter__(self) -> list[Overlap]:\n",
    "        '''\n",
    "        return list of overlaps corresponding to the tree \n",
    "        '''\n",
    "        \n",
    "        frag_list = [self.head]\n",
    "        \n",
    "        if self.tail is not None:\n",
    "            frag_list.extend(list(self.tail)[::-1])\n",
    "            \n",
    "        return iter(frag_list[::-1])\n",
    "\n",
    "class PossibleAssemblies(list):\n",
    "    def __init__(self, sample: str):\n",
    "        self.sample = sample\n",
    "        self.overlap_lls = []\n",
    "        super().__init__(self.overlap_lls)\n",
    "\n",
    "    def update(self, new_overlaps: list[Overlap]):\n",
    "\n",
    "        #If no present linked lists, create them\n",
    "        if not self.overlap_lls:\n",
    "            for overlap in new_overlaps:\n",
    "                self.overlap_lls.append(OverlapLinkedList(overlap))\n",
    "            super().__init__(self.overlap_lls)\n",
    "        #Else, assign them\n",
    "        else:\n",
    "            new_overlap_lls = []\n",
    "            for ll, overlap in product(self.overlap_lls, new_overlaps): \n",
    "                if ll.head.frag_right is overlap.frag_left:\n",
    "                    new_overlap_lls.append(ll.update(overlap))\n",
    "\n",
    "            self.overlap_lls = new_overlap_lls \n",
    "            super().__init__(self.overlap_lls)\n",
    "\n",
    "    def get_final_assemblies(self):\n",
    "\n",
    "        unique_masked_sequences = {}\n",
    "        \n",
    "        for ll in self.overlap_lls:\n",
    "            overlaps = list(ll)\n",
    "\n",
    "            frags = [\n",
    "                overlaps[0].frag_left, \n",
    "                *[\n",
    "                    overlap.frag_right\n",
    "                    for overlap\n",
    "                    in overlaps\n",
    "                ]\n",
    "            ]\n",
    "            unique_masked_sequences.setdefault(tuple(\n",
    "                f.ali.seq_rr_masked\n",
    "                for f\n",
    "                in frags\n",
    "            ), []).append(overlaps)\n",
    "\n",
    "        #print(len(unique_masked_sequences), 'final assembly')\n",
    "        #print('\\n\\n')\n",
    "\n",
    "        if len(unique_masked_sequences) > 2:\n",
    "            raise AssemblyError(\n",
    "                'More than 2 possible sequences - unable to tile across frags',\n",
    "                self.sample,\n",
    "                self.overlap_lls\n",
    "            )\n",
    "        elif len(unique_masked_sequences) == 1:\n",
    "            possibilities = next(iter(unique_masked_sequences.values()))\n",
    "            chosen_overlaps1, chosen_overlaps2 = self.select_two_alleles(\n",
    "                possibilities,\n",
    "                possibilities\n",
    "            )\n",
    "        else:\n",
    "            possibilities = [\n",
    "                ps\n",
    "                for ps\n",
    "                in unique_masked_sequences.values()\n",
    "            ]\n",
    "            chosen_overlaps1, chosen_overlaps2 = self.select_two_alleles(\n",
    "                *possibilities\n",
    "            )\n",
    "\n",
    "        return chosen_overlaps1, chosen_overlaps2\n",
    "\n",
    "    def select_two_alleles(\n",
    "        self,\n",
    "        poss1: list[list[Overlap]], \n",
    "        poss2: list[list[Overlap]]\n",
    "    ) -> tuple[list[Overlap], list[Overlap]]:\n",
    "\n",
    "        acceptable_pair_scores = {}\n",
    "        #Find best combination of overlaps that only consumes each frag once\n",
    "        for overlaps1, overlaps2 in product(poss1, poss2):\n",
    "            #Skip if any overlapping fragments\n",
    "            fragment_ids_1 = {\n",
    "                id(frag)\n",
    "                for overlap\n",
    "                in overlaps1\n",
    "                for frag\n",
    "                in (overlap.frag_left, overlap.frag_right)\n",
    "            }\n",
    "            fragment_ids_2 = {\n",
    "                id(frag)\n",
    "                for overlap\n",
    "                in overlaps2\n",
    "                for frag\n",
    "                in (overlap.frag_left, overlap.frag_right)\n",
    "            }\n",
    "            if fragment_ids_1 & fragment_ids_2:\n",
    "                continue\n",
    "            acceptable_pair_scores[(tuple(overlaps1), tuple(overlaps2))] = sum([\n",
    "                overlap.match_tuple[3] #Sort on kmer distance\n",
    "                for overlaps\n",
    "                in (overlaps1, overlaps2)\n",
    "                for overlap\n",
    "                in overlaps\n",
    "            ])\n",
    "\n",
    "        #Get selected overlaos\n",
    "        overlaps1, overlaps2 = sorted(\n",
    "            acceptable_pair_scores.items(),\n",
    "            key=lambda x: x[1],\n",
    "            reverse=True\n",
    "        )[0][0]\n",
    "\n",
    "        return overlaps1, overlaps2\n",
    "\n",
    "        frags1 = [\n",
    "            overlaps1[0].frag_left, \n",
    "            *[\n",
    "                overlap.frag_right\n",
    "                for overlap\n",
    "                in overlaps1\n",
    "            ]\n",
    "        ]\n",
    "        frags2 = [\n",
    "            overlaps2[0].frag_left, \n",
    "            *[\n",
    "                overlap.frag_right\n",
    "                for overlap\n",
    "                in overlaps2\n",
    "            ]\n",
    "        ]\n",
    "        #for frags in [frags1, frags2]:\n",
    "        #    for frag in frags:\n",
    "        #        print(id(frag), end='\\t')\n",
    "        #    print()\n",
    "        #print(frags1)\n",
    "        #print(frags2)\n",
    "        #print('\\n\\n\\n')\n",
    "        \n",
    "        \n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "        return frags1, frags2\n",
    "\n",
    "\n",
    "                    \n",
    "\n",
    "            \n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48eebbc3-944b-4036-b743-e07075d9a5d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b22f5a5b-1284-45a8-a8cf-55324dacea16",
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = \"out/fails\"\n",
    "\n",
    "# Creates directory and intermediate folders if they don't exist\n",
    "os.makedirs(directory, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1ee7a2d5-3891-4bd4-ad8b-ff60182b9362",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load into sample dict\n",
    "samples = {}\n",
    "\n",
    "for row in rows:\n",
    "    sample = samples.setdefault(row['Samples'], {})\n",
    "\n",
    "    sequence = row['Sequence'].strip()\n",
    "    alignment = piped_read_alns[sequences2i[sequence]]\n",
    "    i_fragment = row['Fragment']\n",
    "    read_count = int(row['NumReads'])\n",
    "    fragment_name = row['Fragment Name']\n",
    "    gen_mms = [\n",
    "        mm\n",
    "        for mm\n",
    "        in (row['gDNA_mismatch_list'].split(', ') if row['gDNA_mismatch_list'] else [])\n",
    "        if mm != 'No_MM'\n",
    "    ]\n",
    "\n",
    "    fragment = Fragment(\n",
    "        seq=sequence,\n",
    "        ali=alignment,\n",
    "        fragment=i_fragment,\n",
    "        read_count=read_count,\n",
    "        fragment_name=fragment_name,\n",
    "        gen_mms=gen_mms\n",
    "    )\n",
    "\n",
    "    if row['analysis_code'] == 1:\n",
    "        sample.setdefault(i_fragment, []).append(fragment)\n",
    "    elif row['analysis_code'] == 2:\n",
    "        sample.setdefault(i_fragment, []).extend([fragment, deepcopy(fragment)])\n",
    "    elif row['analysis_code'] == None:\n",
    "        continue\n",
    "    else:\n",
    "        continue\n",
    "        raise Exception('Other code found:', row['Typing Results::analysis_code'])\n",
    "        \n",
    "    #print(len(sample.get(row['Typing Results::Fragment'], None)))\n",
    "    \n",
    "#Lose weird numbered samples\n",
    "for sample_id, fragment_info in [*samples.items()]:\n",
    "    for i in range(1, 6):\n",
    "    #for fragment, sequences in fragment_info.items():\n",
    "        if len(fragment_info.get(i, [])) != 2:\n",
    "            print('Deleting', sample_id)\n",
    "            print(sample_id.ljust(20), f'fragment {i}', f'{len(fragment_info.get(i, []))} sequences', sep='\\t')\n",
    "            del samples[sample_id]\n",
    "            break\n",
    "\n",
    "    \n",
    "#Lose None sequence\n",
    "for sample_id, fragment_info in [*samples.items()]:\n",
    "    for fragment, sequences in fragment_info.items():\n",
    "        if None in sequences:\n",
    "            print('Deleting', sample_id)\n",
    "            print(sample_id.ljust(20), f'fragment {fragment}', f'{len(sequences)} sequences', sep='\\t')\n",
    "            del samples[sample_id]\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2f90a855-5032-46bd-b5b9-79e6a8af3290",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sample_fragment_pairs = {}\n",
    "\n",
    "for sample_id in samples:\n",
    "    sample_fragment_pairs[sample_id] = []\n",
    "    for i_fragment, fragments in samples[sample_id].items():\n",
    "        fragment_pair = FragmentPair(*fragments, i_fragment)\n",
    "        sample_fragment_pairs[sample_id].append(fragment_pair)\n",
    "    sample_fragment_pairs[sample_id].sort(\n",
    "        key = lambda x: x.fragment\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "95771b17-2a87-47d7-9d60-6ee7b1384e6d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tilings = []\n",
    "for sample_id, fragment_pairs in sample_fragment_pairs.items():\n",
    "    continue\n",
    "    if sample_id != 'DEU':\n",
    "        continue\n",
    "    print(sample_id)\n",
    "    tiling = Tiling.try_init(sample_id, *fragment_pairs)\n",
    "    tilings.append(tiling)\n",
    "    print(tiling.rows)\n",
    "    print('\\n\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "52b4ee68-8eda-4799-b4e2-e51d0870e755",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for tiling in tilings:\n",
    "    for row in tiling.rows:\n",
    "        print(row.tuple)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fcddd7af-2c8b-4c3a-a73c-92e09be3e21b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:   0%|                                                                                                                | 0/66 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.OverlapError'>: No overlaps found between F1 and F2\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:   2%|                                                                                                      | 1/66 [00:00<00:21,  3.07it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.OverlapError'>: F1 could not overlap both frags to F2\n",
      "\n",
      "<class '__main__.OverlapError'>: F3 could not overlap both frags to F4\n",
      "\n",
      "<class '__main__.OverlapError'>: No overlaps found between F4 and F5"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:   5%|                                                                                                   | 3/66 [00:00<00:14,  4.25it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "<class '__main__.OverlapError'>: F2 could not overlap both frags to F3\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:   8%|                                                                                                | 5/66 [00:01<00:11,  5.39it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.AssemblyError'>: More than 2 possible sequences - unable to tile across frags\n",
      "\n",
      "<class '__main__.AssemblyError'>: More than 2 possible sequences - unable to tile across frags<class '__main__.OverlapError'>: F2 could not overlap to both F3 frags\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  11%|                                                                                             | 7/66 [00:01<00:08,  6.94it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.AssemblyError'>: More than 2 possible sequences - unable to tile across frags\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  14%|                                                                                         | 9/66 [00:01<00:06,  8.92it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.AssemblyError'>: More than 2 possible sequences - unable to tile across frags\n",
      "\n",
      "<class '__main__.AssemblyError'>: More than 2 possible sequences - unable to tile across frags\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  26%|                                                                            | 17/66 [02:41<06:37,  8.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.OverlapError'>: F4 could not overlap both frags to F5\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  33%|                                                                    | 22/66 [02:44<01:19,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.OverlapError'>: F2 could not overlap both frags to F3\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  35%|                                                                   | 23/66 [02:44<00:59,  1.37s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.AssemblyError'>: More than 2 possible sequences - unable to tile across frags\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  38%|                                                                | 25/66 [02:45<00:39,  1.05it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.OverlapError'>: F4 could not overlap both frags to F5\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  41%|                                                            | 27/66 [02:46<00:24,  1.56it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.OverlapError'>: F1 could not overlap both frags to F2\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  47%|                                                      | 31/66 [02:48<00:18,  1.88it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.AssemblyError'>: More than 2 possible sequences - unable to tile across frags\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing:  53%|                                                | 35/66 [02:51<00:19,  1.63it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.OverlapError'>: F1 could not overlap both frags to F2\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing: 100%|| 66/66 [05:15<00:00,  4.79s/it]\n"
     ]
    }
   ],
   "source": [
    "from tgsts.utils.parallel import run_concurrently\n",
    "\n",
    "concurrent_results = run_concurrently(\n",
    "    Tiling.try_init,\n",
    "    [\n",
    "        tuple([sample_id, *fragment_pairs])\n",
    "        for sample_id, fragment_pairs\n",
    "        in sample_fragment_pairs.items()\n",
    "    ],\n",
    "    mode = 'process',\n",
    "    raise_exceptions = False,\n",
    "    max_workers = 30,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1746be44-fa2b-4685-a06f-c0a638b29b50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(66, 0)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(concurrent_results.successful), len(concurrent_results.failed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fc91d090-24a1-4571-8027-ae8dcf0fd457",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.FailedTiling at 0x7f4e591a3410>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concurrent_results.successful[0].result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "19f85f73-acf9-428e-b358-4e958523809a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tilings = [\n",
    "    result.result\n",
    "    for result\n",
    "    in concurrent_results.successful\n",
    "]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "734bea4a-98e9-4fdd-96a0-b1cc43172b80",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = 'out/assembled.csv'\n",
    "\n",
    "import csv\n",
    "\n",
    "with open(output_path, 'w', newline='\\n') as csvfile:\n",
    "    writer = csv.writer(\n",
    "        csvfile, \n",
    "        delimiter=',',\n",
    "        quotechar='\"', \n",
    "        quoting=csv.QUOTE_MINIMAL\n",
    "    )\n",
    "    writer.writerow(TilingRow.header)\n",
    "    for tiling in tilings:\n",
    "        if isinstance(tiling, Tiling):\n",
    "            for row in tiling.rows:\n",
    "                writer.writerow(row.tuple)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d5392f29-240c-40f8-9a22-a973ee8bc15c",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = 'out/unassembled.csv'\n",
    "\n",
    "import csv\n",
    "\n",
    "with open(output_path, 'w', newline='\\n') as csvfile:\n",
    "    writer = csv.writer(\n",
    "        csvfile, \n",
    "        delimiter=',',\n",
    "        quotechar='\"', \n",
    "        quoting=csv.QUOTE_MINIMAL\n",
    "    )\n",
    "    writer.writerow(TilingRow.header)\n",
    "    for tiling in tilings:\n",
    "        if not isinstance(tiling, Tiling):\n",
    "            for row in tiling.rows:\n",
    "                writer.writerow(row.tuple)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f07b8b2-9681-40fe-b7f6-8f3bf3353261",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (3.13 tgsts)",
   "language": "python",
   "name": "tgsenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
